{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f97c1cc",
   "metadata": {},
   "source": [
    "# Documents Similarity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3618ccc2",
   "metadata": {},
   "source": [
    "## Read Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4145dd03",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_name_documents = './Databases/prova/prova50.jsonl'\n",
    "#path_name_documents = './Databases/prova/prova30000.jsonl'\n",
    "#path_name_documents = './Databases/prova/prova2000.jsonl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "07b4bfe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import numpy as np\n",
    "import string\n",
    "\n",
    "def readFile(path_name):\n",
    "    # Load the JSONL file into a list\n",
    "    with open(path_name, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    # Convert each JSON object into a dictionary\n",
    "    dicts = [json.loads(line) for line in lines]\n",
    "\n",
    "    # Convert the dictionaries into arrays and stack them vertically\n",
    "    arrays = np.vstack([np.array(list(d.values())) for d in dicts])\n",
    "\n",
    "    # Convert the arrays into a list of lists\n",
    "    text = arrays.tolist()\n",
    "    \n",
    "    return text\n",
    "\n",
    "documents = readFile(path_name_documents)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4f5b47a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "fa45c87e",
   "metadata": {},
   "source": [
    "## Tokenized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a147620",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import nltk\n",
    "import re\n",
    "import string\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem import PorterStemmer, WordNetLemmatizer\n",
    "\n",
    "\n",
    "stop_words = set(stopwords.words('english'))\n",
    "\n",
    "def stemmingLemming(filtered_tokens):\n",
    "    stemmer = PorterStemmer()\n",
    "    lemmatizer = WordNetLemmatizer()\n",
    "\n",
    "    # Perform stemming or lemmatization on filtered tokens\n",
    "    \n",
    "    filtered_tokens = [lemmatizer.lemmatize(token) for token in filtered_tokens]\n",
    "    filtered_tokens = [stemmer.stem(token) for token in filtered_tokens]\n",
    "\n",
    "    return filtered_tokens\n",
    "    \n",
    " \n",
    "    \n",
    "\n",
    "def tokenize(path_name):\n",
    "    \n",
    "    with open(path_name, \"r\") as f:\n",
    "        data = f.readlines()\n",
    "\n",
    "        # Create an empty list to store the tokenized documents\n",
    "        tokenized_docs = []\n",
    "\n",
    "        # Loop through each line in the JSONL file\n",
    "        for line in data:\n",
    "            # Parse the JSON string into a Python dictionary\n",
    "            doc = json.loads(line)\n",
    "\n",
    "            # Extract the text from the dictionary\n",
    "            text = doc['text']\n",
    "            text = text.lower()  # Convert to lowercase\n",
    "            #text = re.sub(r'\\d+', '', text)  # Remove all numbers\n",
    "            text = text.translate(str.maketrans('', '', string.punctuation))  # Remove all punctuation\n",
    "\n",
    "            # Tokenize the text using NLTK\n",
    "            tokens = word_tokenize(text)\n",
    "            tokensStemLem = stemmingLemming(tokens)\n",
    "\n",
    "            # Add the tokenized document to the list\n",
    "            tokenized_docs.append(tokensStemLem)\n",
    "\n",
    "        # Print the tokenized documents\n",
    "    return tokenized_docs\n",
    "\n",
    "\n",
    "tokenized_docs = tokenize(path_name_documents)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96d0ad69",
   "metadata": {},
   "source": [
    "# Sparse Vectors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43553d58",
   "metadata": {},
   "source": [
    "## TF-IDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a261c20e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "\n",
    "def calculateTFIDF(tokenized_docs):\n",
    "    \n",
    "    vectorizer = TfidfVectorizer()\n",
    "    # Fit and transform the tokenized documents into a TF-IDF matrix\n",
    "    tfidf_matrix = vectorizer.fit_transform([' '.join(doc) for doc in tokenized_docs])\n",
    "\n",
    "    # Get the feature names (tokens)\n",
    "    feature_names = vectorizer.get_feature_names_out()\n",
    "\n",
    "    # Return the TF-IDF matrix and the feature names\n",
    "    return tfidf_matrix, feature_names,vectorizer\n",
    "    \n",
    "        \n",
    "\n",
    "tfidf_matrix_docs, feature_names_docs,vectorizer  = calculateTFIDF(tokenized_docs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ce68baf",
   "metadata": {},
   "source": [
    "## Cosine Similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b11ef999",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Doc 1</th>\n",
       "      <th>Doc 2</th>\n",
       "      <th>Doc 3</th>\n",
       "      <th>Doc 4</th>\n",
       "      <th>Doc 5</th>\n",
       "      <th>Doc 6</th>\n",
       "      <th>Doc 7</th>\n",
       "      <th>Doc 8</th>\n",
       "      <th>Doc 9</th>\n",
       "      <th>Doc 10</th>\n",
       "      <th>...</th>\n",
       "      <th>Doc 50</th>\n",
       "      <th>Doc 51</th>\n",
       "      <th>Doc 52</th>\n",
       "      <th>Doc 53</th>\n",
       "      <th>Doc 54</th>\n",
       "      <th>Doc 55</th>\n",
       "      <th>Doc 56</th>\n",
       "      <th>Doc 57</th>\n",
       "      <th>Doc 58</th>\n",
       "      <th>Doc 59</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Doc 1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.116785</td>\n",
       "      <td>0.116682</td>\n",
       "      <td>0.065408</td>\n",
       "      <td>0.185082</td>\n",
       "      <td>0.063252</td>\n",
       "      <td>0.114855</td>\n",
       "      <td>0.131666</td>\n",
       "      <td>0.082319</td>\n",
       "      <td>0.103330</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158555</td>\n",
       "      <td>0.113504</td>\n",
       "      <td>0.183942</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.108963</td>\n",
       "      <td>0.101689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 2</th>\n",
       "      <td>0.116785</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.293539</td>\n",
       "      <td>0.222886</td>\n",
       "      <td>0.172103</td>\n",
       "      <td>0.138513</td>\n",
       "      <td>0.112282</td>\n",
       "      <td>0.165781</td>\n",
       "      <td>0.222440</td>\n",
       "      <td>0.171459</td>\n",
       "      <td>...</td>\n",
       "      <td>0.258898</td>\n",
       "      <td>0.114252</td>\n",
       "      <td>0.159238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.161876</td>\n",
       "      <td>0.175653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 3</th>\n",
       "      <td>0.116682</td>\n",
       "      <td>0.293539</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.181477</td>\n",
       "      <td>0.235975</td>\n",
       "      <td>0.105975</td>\n",
       "      <td>0.114559</td>\n",
       "      <td>0.155222</td>\n",
       "      <td>0.172260</td>\n",
       "      <td>0.145168</td>\n",
       "      <td>...</td>\n",
       "      <td>0.233060</td>\n",
       "      <td>0.160828</td>\n",
       "      <td>0.175988</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.185843</td>\n",
       "      <td>0.185424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 4</th>\n",
       "      <td>0.065408</td>\n",
       "      <td>0.222886</td>\n",
       "      <td>0.181477</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.097654</td>\n",
       "      <td>0.063610</td>\n",
       "      <td>0.063011</td>\n",
       "      <td>0.093672</td>\n",
       "      <td>0.162503</td>\n",
       "      <td>0.115923</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161168</td>\n",
       "      <td>0.089334</td>\n",
       "      <td>0.118018</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.094309</td>\n",
       "      <td>0.136627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 5</th>\n",
       "      <td>0.185082</td>\n",
       "      <td>0.172103</td>\n",
       "      <td>0.235975</td>\n",
       "      <td>0.097654</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.096868</td>\n",
       "      <td>0.139804</td>\n",
       "      <td>0.132900</td>\n",
       "      <td>0.149990</td>\n",
       "      <td>0.128149</td>\n",
       "      <td>...</td>\n",
       "      <td>0.193689</td>\n",
       "      <td>0.128157</td>\n",
       "      <td>0.174898</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.179021</td>\n",
       "      <td>0.182024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 6</th>\n",
       "      <td>0.063252</td>\n",
       "      <td>0.138513</td>\n",
       "      <td>0.105975</td>\n",
       "      <td>0.063610</td>\n",
       "      <td>0.096868</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.046924</td>\n",
       "      <td>0.094784</td>\n",
       "      <td>0.082026</td>\n",
       "      <td>0.094177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.141220</td>\n",
       "      <td>0.067741</td>\n",
       "      <td>0.098841</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125477</td>\n",
       "      <td>0.111974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 7</th>\n",
       "      <td>0.114855</td>\n",
       "      <td>0.112282</td>\n",
       "      <td>0.114559</td>\n",
       "      <td>0.063011</td>\n",
       "      <td>0.139804</td>\n",
       "      <td>0.046924</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.135525</td>\n",
       "      <td>0.097209</td>\n",
       "      <td>0.099001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.184630</td>\n",
       "      <td>0.084729</td>\n",
       "      <td>0.173973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.090519</td>\n",
       "      <td>0.081340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 8</th>\n",
       "      <td>0.131666</td>\n",
       "      <td>0.165781</td>\n",
       "      <td>0.155222</td>\n",
       "      <td>0.093672</td>\n",
       "      <td>0.132900</td>\n",
       "      <td>0.094784</td>\n",
       "      <td>0.135525</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.109231</td>\n",
       "      <td>0.147724</td>\n",
       "      <td>...</td>\n",
       "      <td>0.271639</td>\n",
       "      <td>0.089889</td>\n",
       "      <td>0.125835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.152414</td>\n",
       "      <td>0.124312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 9</th>\n",
       "      <td>0.082319</td>\n",
       "      <td>0.222440</td>\n",
       "      <td>0.172260</td>\n",
       "      <td>0.162503</td>\n",
       "      <td>0.149990</td>\n",
       "      <td>0.082026</td>\n",
       "      <td>0.097209</td>\n",
       "      <td>0.109231</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.118235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.165711</td>\n",
       "      <td>0.099433</td>\n",
       "      <td>0.145940</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.139159</td>\n",
       "      <td>0.140329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 10</th>\n",
       "      <td>0.103330</td>\n",
       "      <td>0.171459</td>\n",
       "      <td>0.145168</td>\n",
       "      <td>0.115923</td>\n",
       "      <td>0.128149</td>\n",
       "      <td>0.094177</td>\n",
       "      <td>0.099001</td>\n",
       "      <td>0.147724</td>\n",
       "      <td>0.118235</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.236668</td>\n",
       "      <td>0.102157</td>\n",
       "      <td>0.174861</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.165920</td>\n",
       "      <td>0.165466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 11</th>\n",
       "      <td>0.099578</td>\n",
       "      <td>0.134399</td>\n",
       "      <td>0.142115</td>\n",
       "      <td>0.070597</td>\n",
       "      <td>0.142067</td>\n",
       "      <td>0.113620</td>\n",
       "      <td>0.059800</td>\n",
       "      <td>0.097432</td>\n",
       "      <td>0.096080</td>\n",
       "      <td>0.116509</td>\n",
       "      <td>...</td>\n",
       "      <td>0.150959</td>\n",
       "      <td>0.060228</td>\n",
       "      <td>0.150303</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.174335</td>\n",
       "      <td>0.112852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 12</th>\n",
       "      <td>0.073413</td>\n",
       "      <td>0.137236</td>\n",
       "      <td>0.149265</td>\n",
       "      <td>0.073216</td>\n",
       "      <td>0.137934</td>\n",
       "      <td>0.186082</td>\n",
       "      <td>0.063900</td>\n",
       "      <td>0.108100</td>\n",
       "      <td>0.113032</td>\n",
       "      <td>0.110699</td>\n",
       "      <td>...</td>\n",
       "      <td>0.148205</td>\n",
       "      <td>0.075926</td>\n",
       "      <td>0.109810</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.148582</td>\n",
       "      <td>0.120566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 13</th>\n",
       "      <td>0.069734</td>\n",
       "      <td>0.124156</td>\n",
       "      <td>0.121149</td>\n",
       "      <td>0.105220</td>\n",
       "      <td>0.104601</td>\n",
       "      <td>0.085342</td>\n",
       "      <td>0.051863</td>\n",
       "      <td>0.073111</td>\n",
       "      <td>0.103513</td>\n",
       "      <td>0.097210</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131324</td>\n",
       "      <td>0.054946</td>\n",
       "      <td>0.091295</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.135736</td>\n",
       "      <td>0.106503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 14</th>\n",
       "      <td>0.104889</td>\n",
       "      <td>0.160809</td>\n",
       "      <td>0.163465</td>\n",
       "      <td>0.109380</td>\n",
       "      <td>0.173094</td>\n",
       "      <td>0.107368</td>\n",
       "      <td>0.075794</td>\n",
       "      <td>0.110996</td>\n",
       "      <td>0.121068</td>\n",
       "      <td>0.116875</td>\n",
       "      <td>...</td>\n",
       "      <td>0.186318</td>\n",
       "      <td>0.084827</td>\n",
       "      <td>0.140457</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.138115</td>\n",
       "      <td>0.155765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 15</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 16</th>\n",
       "      <td>0.164147</td>\n",
       "      <td>0.183739</td>\n",
       "      <td>0.206926</td>\n",
       "      <td>0.144194</td>\n",
       "      <td>0.330745</td>\n",
       "      <td>0.118418</td>\n",
       "      <td>0.148003</td>\n",
       "      <td>0.166809</td>\n",
       "      <td>0.159087</td>\n",
       "      <td>0.211481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.239887</td>\n",
       "      <td>0.150346</td>\n",
       "      <td>0.207647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.209576</td>\n",
       "      <td>0.256975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 17</th>\n",
       "      <td>0.195453</td>\n",
       "      <td>0.182371</td>\n",
       "      <td>0.188846</td>\n",
       "      <td>0.134475</td>\n",
       "      <td>0.159141</td>\n",
       "      <td>0.103519</td>\n",
       "      <td>0.208224</td>\n",
       "      <td>0.142852</td>\n",
       "      <td>0.140772</td>\n",
       "      <td>0.136021</td>\n",
       "      <td>...</td>\n",
       "      <td>0.192152</td>\n",
       "      <td>0.114266</td>\n",
       "      <td>0.184255</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.160721</td>\n",
       "      <td>0.125082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 18</th>\n",
       "      <td>0.059993</td>\n",
       "      <td>0.117804</td>\n",
       "      <td>0.105060</td>\n",
       "      <td>0.054958</td>\n",
       "      <td>0.099563</td>\n",
       "      <td>0.057566</td>\n",
       "      <td>0.064901</td>\n",
       "      <td>0.066223</td>\n",
       "      <td>0.083643</td>\n",
       "      <td>0.097712</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131048</td>\n",
       "      <td>0.062794</td>\n",
       "      <td>0.105670</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.077557</td>\n",
       "      <td>0.125615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 19</th>\n",
       "      <td>0.095529</td>\n",
       "      <td>0.140455</td>\n",
       "      <td>0.187573</td>\n",
       "      <td>0.070020</td>\n",
       "      <td>0.139712</td>\n",
       "      <td>0.070413</td>\n",
       "      <td>0.101842</td>\n",
       "      <td>0.094137</td>\n",
       "      <td>0.117368</td>\n",
       "      <td>0.107373</td>\n",
       "      <td>...</td>\n",
       "      <td>0.135323</td>\n",
       "      <td>0.167085</td>\n",
       "      <td>0.132014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.175687</td>\n",
       "      <td>0.121154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 20</th>\n",
       "      <td>0.108649</td>\n",
       "      <td>0.185489</td>\n",
       "      <td>0.149524</td>\n",
       "      <td>0.102986</td>\n",
       "      <td>0.194331</td>\n",
       "      <td>0.149575</td>\n",
       "      <td>0.086304</td>\n",
       "      <td>0.105564</td>\n",
       "      <td>0.156089</td>\n",
       "      <td>0.148262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.178179</td>\n",
       "      <td>0.085107</td>\n",
       "      <td>0.149336</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.177183</td>\n",
       "      <td>0.181017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 21</th>\n",
       "      <td>0.085536</td>\n",
       "      <td>0.163982</td>\n",
       "      <td>0.154307</td>\n",
       "      <td>0.129735</td>\n",
       "      <td>0.114131</td>\n",
       "      <td>0.077863</td>\n",
       "      <td>0.076457</td>\n",
       "      <td>0.159945</td>\n",
       "      <td>0.132522</td>\n",
       "      <td>0.181965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.239480</td>\n",
       "      <td>0.102407</td>\n",
       "      <td>0.152915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.142773</td>\n",
       "      <td>0.170131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 22</th>\n",
       "      <td>0.087727</td>\n",
       "      <td>0.169946</td>\n",
       "      <td>0.168107</td>\n",
       "      <td>0.103071</td>\n",
       "      <td>0.137439</td>\n",
       "      <td>0.095326</td>\n",
       "      <td>0.080486</td>\n",
       "      <td>0.111076</td>\n",
       "      <td>0.140537</td>\n",
       "      <td>0.145046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.160755</td>\n",
       "      <td>0.074856</td>\n",
       "      <td>0.162850</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.136043</td>\n",
       "      <td>0.119413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 23</th>\n",
       "      <td>0.099840</td>\n",
       "      <td>0.178411</td>\n",
       "      <td>0.167336</td>\n",
       "      <td>0.114519</td>\n",
       "      <td>0.144830</td>\n",
       "      <td>0.162783</td>\n",
       "      <td>0.078852</td>\n",
       "      <td>0.125713</td>\n",
       "      <td>0.138715</td>\n",
       "      <td>0.170944</td>\n",
       "      <td>...</td>\n",
       "      <td>0.194180</td>\n",
       "      <td>0.100097</td>\n",
       "      <td>0.155831</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.204906</td>\n",
       "      <td>0.192122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 24</th>\n",
       "      <td>0.083684</td>\n",
       "      <td>0.137663</td>\n",
       "      <td>0.154619</td>\n",
       "      <td>0.064416</td>\n",
       "      <td>0.142416</td>\n",
       "      <td>0.091920</td>\n",
       "      <td>0.081709</td>\n",
       "      <td>0.098792</td>\n",
       "      <td>0.104502</td>\n",
       "      <td>0.134515</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156082</td>\n",
       "      <td>0.081920</td>\n",
       "      <td>0.143701</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.186991</td>\n",
       "      <td>0.140449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 25</th>\n",
       "      <td>0.097007</td>\n",
       "      <td>0.153771</td>\n",
       "      <td>0.172393</td>\n",
       "      <td>0.102027</td>\n",
       "      <td>0.145962</td>\n",
       "      <td>0.087208</td>\n",
       "      <td>0.089324</td>\n",
       "      <td>0.128907</td>\n",
       "      <td>0.116363</td>\n",
       "      <td>0.117308</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166920</td>\n",
       "      <td>0.116938</td>\n",
       "      <td>0.130530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.176602</td>\n",
       "      <td>0.127690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 26</th>\n",
       "      <td>0.093745</td>\n",
       "      <td>0.092828</td>\n",
       "      <td>0.117982</td>\n",
       "      <td>0.133232</td>\n",
       "      <td>0.105276</td>\n",
       "      <td>0.049403</td>\n",
       "      <td>0.060495</td>\n",
       "      <td>0.074547</td>\n",
       "      <td>0.103582</td>\n",
       "      <td>0.070872</td>\n",
       "      <td>...</td>\n",
       "      <td>0.118898</td>\n",
       "      <td>0.125293</td>\n",
       "      <td>0.123362</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.100963</td>\n",
       "      <td>0.090770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 27</th>\n",
       "      <td>0.132481</td>\n",
       "      <td>0.164923</td>\n",
       "      <td>0.138406</td>\n",
       "      <td>0.107310</td>\n",
       "      <td>0.194255</td>\n",
       "      <td>0.159590</td>\n",
       "      <td>0.071737</td>\n",
       "      <td>0.128998</td>\n",
       "      <td>0.094528</td>\n",
       "      <td>0.136686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.180698</td>\n",
       "      <td>0.100513</td>\n",
       "      <td>0.163864</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.146929</td>\n",
       "      <td>0.182905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 28</th>\n",
       "      <td>0.086362</td>\n",
       "      <td>0.160953</td>\n",
       "      <td>0.129713</td>\n",
       "      <td>0.090320</td>\n",
       "      <td>0.093763</td>\n",
       "      <td>0.171876</td>\n",
       "      <td>0.084680</td>\n",
       "      <td>0.116039</td>\n",
       "      <td>0.105223</td>\n",
       "      <td>0.113288</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166193</td>\n",
       "      <td>0.067261</td>\n",
       "      <td>0.133989</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.150759</td>\n",
       "      <td>0.136666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 29</th>\n",
       "      <td>0.093049</td>\n",
       "      <td>0.146017</td>\n",
       "      <td>0.141014</td>\n",
       "      <td>0.062329</td>\n",
       "      <td>0.125561</td>\n",
       "      <td>0.119790</td>\n",
       "      <td>0.104170</td>\n",
       "      <td>0.100485</td>\n",
       "      <td>0.096546</td>\n",
       "      <td>0.107054</td>\n",
       "      <td>...</td>\n",
       "      <td>0.164877</td>\n",
       "      <td>0.082074</td>\n",
       "      <td>0.125202</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.138711</td>\n",
       "      <td>0.129555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 30</th>\n",
       "      <td>0.115417</td>\n",
       "      <td>0.201283</td>\n",
       "      <td>0.201626</td>\n",
       "      <td>0.087839</td>\n",
       "      <td>0.174249</td>\n",
       "      <td>0.194022</td>\n",
       "      <td>0.073656</td>\n",
       "      <td>0.160880</td>\n",
       "      <td>0.129961</td>\n",
       "      <td>0.206259</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222120</td>\n",
       "      <td>0.103976</td>\n",
       "      <td>0.183047</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.209475</td>\n",
       "      <td>0.161882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 31</th>\n",
       "      <td>0.106411</td>\n",
       "      <td>0.160290</td>\n",
       "      <td>0.142353</td>\n",
       "      <td>0.097822</td>\n",
       "      <td>0.117680</td>\n",
       "      <td>0.124222</td>\n",
       "      <td>0.080784</td>\n",
       "      <td>0.104529</td>\n",
       "      <td>0.120069</td>\n",
       "      <td>0.122733</td>\n",
       "      <td>...</td>\n",
       "      <td>0.163592</td>\n",
       "      <td>0.097858</td>\n",
       "      <td>0.143472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.151850</td>\n",
       "      <td>0.102236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 32</th>\n",
       "      <td>0.110149</td>\n",
       "      <td>0.156088</td>\n",
       "      <td>0.149881</td>\n",
       "      <td>0.089335</td>\n",
       "      <td>0.113868</td>\n",
       "      <td>0.091946</td>\n",
       "      <td>0.112937</td>\n",
       "      <td>0.141983</td>\n",
       "      <td>0.129830</td>\n",
       "      <td>0.235001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.290518</td>\n",
       "      <td>0.100291</td>\n",
       "      <td>0.182947</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170552</td>\n",
       "      <td>0.152193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 33</th>\n",
       "      <td>0.107940</td>\n",
       "      <td>0.203815</td>\n",
       "      <td>0.174205</td>\n",
       "      <td>0.164322</td>\n",
       "      <td>0.131662</td>\n",
       "      <td>0.094116</td>\n",
       "      <td>0.093187</td>\n",
       "      <td>0.168514</td>\n",
       "      <td>0.131007</td>\n",
       "      <td>0.207040</td>\n",
       "      <td>...</td>\n",
       "      <td>0.204532</td>\n",
       "      <td>0.104100</td>\n",
       "      <td>0.172836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.160454</td>\n",
       "      <td>0.159194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 34</th>\n",
       "      <td>0.099877</td>\n",
       "      <td>0.217013</td>\n",
       "      <td>0.208455</td>\n",
       "      <td>0.144328</td>\n",
       "      <td>0.151802</td>\n",
       "      <td>0.099520</td>\n",
       "      <td>0.120845</td>\n",
       "      <td>0.133625</td>\n",
       "      <td>0.157121</td>\n",
       "      <td>0.216711</td>\n",
       "      <td>...</td>\n",
       "      <td>0.296036</td>\n",
       "      <td>0.113282</td>\n",
       "      <td>0.213479</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.158022</td>\n",
       "      <td>0.182414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 35</th>\n",
       "      <td>0.099141</td>\n",
       "      <td>0.171135</td>\n",
       "      <td>0.141984</td>\n",
       "      <td>0.111260</td>\n",
       "      <td>0.126550</td>\n",
       "      <td>0.110368</td>\n",
       "      <td>0.083056</td>\n",
       "      <td>0.131145</td>\n",
       "      <td>0.110865</td>\n",
       "      <td>0.115268</td>\n",
       "      <td>...</td>\n",
       "      <td>0.174731</td>\n",
       "      <td>0.086845</td>\n",
       "      <td>0.133350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.140525</td>\n",
       "      <td>0.123981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 36</th>\n",
       "      <td>0.096055</td>\n",
       "      <td>0.235732</td>\n",
       "      <td>0.175612</td>\n",
       "      <td>0.122778</td>\n",
       "      <td>0.182883</td>\n",
       "      <td>0.272676</td>\n",
       "      <td>0.079249</td>\n",
       "      <td>0.111982</td>\n",
       "      <td>0.129016</td>\n",
       "      <td>0.197003</td>\n",
       "      <td>...</td>\n",
       "      <td>0.214439</td>\n",
       "      <td>0.095076</td>\n",
       "      <td>0.164451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.172902</td>\n",
       "      <td>0.206983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 37</th>\n",
       "      <td>0.128216</td>\n",
       "      <td>0.174025</td>\n",
       "      <td>0.190269</td>\n",
       "      <td>0.132389</td>\n",
       "      <td>0.183267</td>\n",
       "      <td>0.094621</td>\n",
       "      <td>0.111549</td>\n",
       "      <td>0.138726</td>\n",
       "      <td>0.138802</td>\n",
       "      <td>0.122547</td>\n",
       "      <td>...</td>\n",
       "      <td>0.186266</td>\n",
       "      <td>0.135514</td>\n",
       "      <td>0.139725</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.170663</td>\n",
       "      <td>0.127123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 38</th>\n",
       "      <td>0.058383</td>\n",
       "      <td>0.097927</td>\n",
       "      <td>0.090741</td>\n",
       "      <td>0.043953</td>\n",
       "      <td>0.072172</td>\n",
       "      <td>0.101332</td>\n",
       "      <td>0.044520</td>\n",
       "      <td>0.068699</td>\n",
       "      <td>0.063132</td>\n",
       "      <td>0.069004</td>\n",
       "      <td>...</td>\n",
       "      <td>0.101152</td>\n",
       "      <td>0.062345</td>\n",
       "      <td>0.087919</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.081703</td>\n",
       "      <td>0.071785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 39</th>\n",
       "      <td>0.065006</td>\n",
       "      <td>0.127444</td>\n",
       "      <td>0.127185</td>\n",
       "      <td>0.066576</td>\n",
       "      <td>0.122151</td>\n",
       "      <td>0.140995</td>\n",
       "      <td>0.072888</td>\n",
       "      <td>0.090382</td>\n",
       "      <td>0.082448</td>\n",
       "      <td>0.095109</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158367</td>\n",
       "      <td>0.067036</td>\n",
       "      <td>0.112057</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.131828</td>\n",
       "      <td>0.153358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 40</th>\n",
       "      <td>0.105027</td>\n",
       "      <td>0.181750</td>\n",
       "      <td>0.221176</td>\n",
       "      <td>0.102002</td>\n",
       "      <td>0.189477</td>\n",
       "      <td>0.106387</td>\n",
       "      <td>0.074531</td>\n",
       "      <td>0.112792</td>\n",
       "      <td>0.117850</td>\n",
       "      <td>0.175680</td>\n",
       "      <td>...</td>\n",
       "      <td>0.202966</td>\n",
       "      <td>0.118186</td>\n",
       "      <td>0.162316</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.137983</td>\n",
       "      <td>0.211101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 41</th>\n",
       "      <td>0.124855</td>\n",
       "      <td>0.149154</td>\n",
       "      <td>0.147800</td>\n",
       "      <td>0.092835</td>\n",
       "      <td>0.174120</td>\n",
       "      <td>0.097018</td>\n",
       "      <td>0.092848</td>\n",
       "      <td>0.140819</td>\n",
       "      <td>0.127682</td>\n",
       "      <td>0.148133</td>\n",
       "      <td>...</td>\n",
       "      <td>0.176488</td>\n",
       "      <td>0.108562</td>\n",
       "      <td>0.173651</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.261871</td>\n",
       "      <td>0.116705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 42</th>\n",
       "      <td>0.026485</td>\n",
       "      <td>0.070644</td>\n",
       "      <td>0.059525</td>\n",
       "      <td>0.064289</td>\n",
       "      <td>0.104305</td>\n",
       "      <td>0.052123</td>\n",
       "      <td>0.060025</td>\n",
       "      <td>0.029710</td>\n",
       "      <td>0.058840</td>\n",
       "      <td>0.121969</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166598</td>\n",
       "      <td>0.051721</td>\n",
       "      <td>0.075256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.046876</td>\n",
       "      <td>0.055075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 43</th>\n",
       "      <td>0.230433</td>\n",
       "      <td>0.113461</td>\n",
       "      <td>0.114646</td>\n",
       "      <td>0.075539</td>\n",
       "      <td>0.131419</td>\n",
       "      <td>0.042823</td>\n",
       "      <td>0.110839</td>\n",
       "      <td>0.092156</td>\n",
       "      <td>0.091469</td>\n",
       "      <td>0.085491</td>\n",
       "      <td>...</td>\n",
       "      <td>0.129270</td>\n",
       "      <td>0.097456</td>\n",
       "      <td>0.261086</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091827</td>\n",
       "      <td>0.098842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 44</th>\n",
       "      <td>0.059066</td>\n",
       "      <td>0.086947</td>\n",
       "      <td>0.083681</td>\n",
       "      <td>0.040270</td>\n",
       "      <td>0.058279</td>\n",
       "      <td>0.056986</td>\n",
       "      <td>0.037105</td>\n",
       "      <td>0.108924</td>\n",
       "      <td>0.053782</td>\n",
       "      <td>0.110771</td>\n",
       "      <td>...</td>\n",
       "      <td>0.172829</td>\n",
       "      <td>0.054417</td>\n",
       "      <td>0.056506</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088176</td>\n",
       "      <td>0.076290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 45</th>\n",
       "      <td>0.137321</td>\n",
       "      <td>0.156401</td>\n",
       "      <td>0.155912</td>\n",
       "      <td>0.077321</td>\n",
       "      <td>0.138349</td>\n",
       "      <td>0.085877</td>\n",
       "      <td>0.099268</td>\n",
       "      <td>0.145881</td>\n",
       "      <td>0.130406</td>\n",
       "      <td>0.108719</td>\n",
       "      <td>...</td>\n",
       "      <td>0.200937</td>\n",
       "      <td>0.166008</td>\n",
       "      <td>0.150386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.147125</td>\n",
       "      <td>0.144709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 46</th>\n",
       "      <td>0.109397</td>\n",
       "      <td>0.169957</td>\n",
       "      <td>0.153174</td>\n",
       "      <td>0.086246</td>\n",
       "      <td>0.138555</td>\n",
       "      <td>0.133969</td>\n",
       "      <td>0.055313</td>\n",
       "      <td>0.146149</td>\n",
       "      <td>0.119740</td>\n",
       "      <td>0.150022</td>\n",
       "      <td>...</td>\n",
       "      <td>0.187400</td>\n",
       "      <td>0.088309</td>\n",
       "      <td>0.138912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.201635</td>\n",
       "      <td>0.133507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 47</th>\n",
       "      <td>0.073457</td>\n",
       "      <td>0.116354</td>\n",
       "      <td>0.135087</td>\n",
       "      <td>0.082397</td>\n",
       "      <td>0.101719</td>\n",
       "      <td>0.060348</td>\n",
       "      <td>0.091893</td>\n",
       "      <td>0.081322</td>\n",
       "      <td>0.098389</td>\n",
       "      <td>0.129807</td>\n",
       "      <td>...</td>\n",
       "      <td>0.140852</td>\n",
       "      <td>0.076504</td>\n",
       "      <td>0.136313</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.099747</td>\n",
       "      <td>0.103475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 48</th>\n",
       "      <td>0.104151</td>\n",
       "      <td>0.111153</td>\n",
       "      <td>0.164448</td>\n",
       "      <td>0.069449</td>\n",
       "      <td>0.146907</td>\n",
       "      <td>0.076324</td>\n",
       "      <td>0.072458</td>\n",
       "      <td>0.086895</td>\n",
       "      <td>0.105628</td>\n",
       "      <td>0.097056</td>\n",
       "      <td>...</td>\n",
       "      <td>0.133022</td>\n",
       "      <td>0.076530</td>\n",
       "      <td>0.111074</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083653</td>\n",
       "      <td>0.119951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 49</th>\n",
       "      <td>0.102567</td>\n",
       "      <td>0.145953</td>\n",
       "      <td>0.164013</td>\n",
       "      <td>0.084601</td>\n",
       "      <td>0.164828</td>\n",
       "      <td>0.098536</td>\n",
       "      <td>0.123826</td>\n",
       "      <td>0.103991</td>\n",
       "      <td>0.137495</td>\n",
       "      <td>0.218288</td>\n",
       "      <td>...</td>\n",
       "      <td>0.325398</td>\n",
       "      <td>0.072165</td>\n",
       "      <td>0.183243</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.129795</td>\n",
       "      <td>0.153425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 50</th>\n",
       "      <td>0.158555</td>\n",
       "      <td>0.258898</td>\n",
       "      <td>0.233060</td>\n",
       "      <td>0.161168</td>\n",
       "      <td>0.193689</td>\n",
       "      <td>0.141220</td>\n",
       "      <td>0.184630</td>\n",
       "      <td>0.271639</td>\n",
       "      <td>0.165711</td>\n",
       "      <td>0.236668</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.130866</td>\n",
       "      <td>0.223219</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.184290</td>\n",
       "      <td>0.209632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 51</th>\n",
       "      <td>0.113504</td>\n",
       "      <td>0.114252</td>\n",
       "      <td>0.160828</td>\n",
       "      <td>0.089334</td>\n",
       "      <td>0.128157</td>\n",
       "      <td>0.067741</td>\n",
       "      <td>0.084729</td>\n",
       "      <td>0.089889</td>\n",
       "      <td>0.099433</td>\n",
       "      <td>0.102157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130866</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.121413</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.127521</td>\n",
       "      <td>0.127359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 52</th>\n",
       "      <td>0.183942</td>\n",
       "      <td>0.159238</td>\n",
       "      <td>0.175988</td>\n",
       "      <td>0.118018</td>\n",
       "      <td>0.174898</td>\n",
       "      <td>0.098841</td>\n",
       "      <td>0.173973</td>\n",
       "      <td>0.125835</td>\n",
       "      <td>0.145940</td>\n",
       "      <td>0.174861</td>\n",
       "      <td>...</td>\n",
       "      <td>0.223219</td>\n",
       "      <td>0.121413</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.162394</td>\n",
       "      <td>0.156033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 53</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 54</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 55</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 56</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 57</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 58</th>\n",
       "      <td>0.108963</td>\n",
       "      <td>0.161876</td>\n",
       "      <td>0.185843</td>\n",
       "      <td>0.094309</td>\n",
       "      <td>0.179021</td>\n",
       "      <td>0.125477</td>\n",
       "      <td>0.090519</td>\n",
       "      <td>0.152414</td>\n",
       "      <td>0.139159</td>\n",
       "      <td>0.165920</td>\n",
       "      <td>...</td>\n",
       "      <td>0.184290</td>\n",
       "      <td>0.127521</td>\n",
       "      <td>0.162394</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.214111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Doc 59</th>\n",
       "      <td>0.101689</td>\n",
       "      <td>0.175653</td>\n",
       "      <td>0.185424</td>\n",
       "      <td>0.136627</td>\n",
       "      <td>0.182024</td>\n",
       "      <td>0.111974</td>\n",
       "      <td>0.081340</td>\n",
       "      <td>0.124312</td>\n",
       "      <td>0.140329</td>\n",
       "      <td>0.165466</td>\n",
       "      <td>...</td>\n",
       "      <td>0.209632</td>\n",
       "      <td>0.127359</td>\n",
       "      <td>0.156033</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.214111</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Doc 1     Doc 2     Doc 3     Doc 4     Doc 5     Doc 6     Doc 7  \\\n",
       "Doc 1   1.000000  0.116785  0.116682  0.065408  0.185082  0.063252  0.114855   \n",
       "Doc 2   0.116785  1.000000  0.293539  0.222886  0.172103  0.138513  0.112282   \n",
       "Doc 3   0.116682  0.293539  1.000000  0.181477  0.235975  0.105975  0.114559   \n",
       "Doc 4   0.065408  0.222886  0.181477  1.000000  0.097654  0.063610  0.063011   \n",
       "Doc 5   0.185082  0.172103  0.235975  0.097654  1.000000  0.096868  0.139804   \n",
       "Doc 6   0.063252  0.138513  0.105975  0.063610  0.096868  1.000000  0.046924   \n",
       "Doc 7   0.114855  0.112282  0.114559  0.063011  0.139804  0.046924  1.000000   \n",
       "Doc 8   0.131666  0.165781  0.155222  0.093672  0.132900  0.094784  0.135525   \n",
       "Doc 9   0.082319  0.222440  0.172260  0.162503  0.149990  0.082026  0.097209   \n",
       "Doc 10  0.103330  0.171459  0.145168  0.115923  0.128149  0.094177  0.099001   \n",
       "Doc 11  0.099578  0.134399  0.142115  0.070597  0.142067  0.113620  0.059800   \n",
       "Doc 12  0.073413  0.137236  0.149265  0.073216  0.137934  0.186082  0.063900   \n",
       "Doc 13  0.069734  0.124156  0.121149  0.105220  0.104601  0.085342  0.051863   \n",
       "Doc 14  0.104889  0.160809  0.163465  0.109380  0.173094  0.107368  0.075794   \n",
       "Doc 15  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 16  0.164147  0.183739  0.206926  0.144194  0.330745  0.118418  0.148003   \n",
       "Doc 17  0.195453  0.182371  0.188846  0.134475  0.159141  0.103519  0.208224   \n",
       "Doc 18  0.059993  0.117804  0.105060  0.054958  0.099563  0.057566  0.064901   \n",
       "Doc 19  0.095529  0.140455  0.187573  0.070020  0.139712  0.070413  0.101842   \n",
       "Doc 20  0.108649  0.185489  0.149524  0.102986  0.194331  0.149575  0.086304   \n",
       "Doc 21  0.085536  0.163982  0.154307  0.129735  0.114131  0.077863  0.076457   \n",
       "Doc 22  0.087727  0.169946  0.168107  0.103071  0.137439  0.095326  0.080486   \n",
       "Doc 23  0.099840  0.178411  0.167336  0.114519  0.144830  0.162783  0.078852   \n",
       "Doc 24  0.083684  0.137663  0.154619  0.064416  0.142416  0.091920  0.081709   \n",
       "Doc 25  0.097007  0.153771  0.172393  0.102027  0.145962  0.087208  0.089324   \n",
       "Doc 26  0.093745  0.092828  0.117982  0.133232  0.105276  0.049403  0.060495   \n",
       "Doc 27  0.132481  0.164923  0.138406  0.107310  0.194255  0.159590  0.071737   \n",
       "Doc 28  0.086362  0.160953  0.129713  0.090320  0.093763  0.171876  0.084680   \n",
       "Doc 29  0.093049  0.146017  0.141014  0.062329  0.125561  0.119790  0.104170   \n",
       "Doc 30  0.115417  0.201283  0.201626  0.087839  0.174249  0.194022  0.073656   \n",
       "Doc 31  0.106411  0.160290  0.142353  0.097822  0.117680  0.124222  0.080784   \n",
       "Doc 32  0.110149  0.156088  0.149881  0.089335  0.113868  0.091946  0.112937   \n",
       "Doc 33  0.107940  0.203815  0.174205  0.164322  0.131662  0.094116  0.093187   \n",
       "Doc 34  0.099877  0.217013  0.208455  0.144328  0.151802  0.099520  0.120845   \n",
       "Doc 35  0.099141  0.171135  0.141984  0.111260  0.126550  0.110368  0.083056   \n",
       "Doc 36  0.096055  0.235732  0.175612  0.122778  0.182883  0.272676  0.079249   \n",
       "Doc 37  0.128216  0.174025  0.190269  0.132389  0.183267  0.094621  0.111549   \n",
       "Doc 38  0.058383  0.097927  0.090741  0.043953  0.072172  0.101332  0.044520   \n",
       "Doc 39  0.065006  0.127444  0.127185  0.066576  0.122151  0.140995  0.072888   \n",
       "Doc 40  0.105027  0.181750  0.221176  0.102002  0.189477  0.106387  0.074531   \n",
       "Doc 41  0.124855  0.149154  0.147800  0.092835  0.174120  0.097018  0.092848   \n",
       "Doc 42  0.026485  0.070644  0.059525  0.064289  0.104305  0.052123  0.060025   \n",
       "Doc 43  0.230433  0.113461  0.114646  0.075539  0.131419  0.042823  0.110839   \n",
       "Doc 44  0.059066  0.086947  0.083681  0.040270  0.058279  0.056986  0.037105   \n",
       "Doc 45  0.137321  0.156401  0.155912  0.077321  0.138349  0.085877  0.099268   \n",
       "Doc 46  0.109397  0.169957  0.153174  0.086246  0.138555  0.133969  0.055313   \n",
       "Doc 47  0.073457  0.116354  0.135087  0.082397  0.101719  0.060348  0.091893   \n",
       "Doc 48  0.104151  0.111153  0.164448  0.069449  0.146907  0.076324  0.072458   \n",
       "Doc 49  0.102567  0.145953  0.164013  0.084601  0.164828  0.098536  0.123826   \n",
       "Doc 50  0.158555  0.258898  0.233060  0.161168  0.193689  0.141220  0.184630   \n",
       "Doc 51  0.113504  0.114252  0.160828  0.089334  0.128157  0.067741  0.084729   \n",
       "Doc 52  0.183942  0.159238  0.175988  0.118018  0.174898  0.098841  0.173973   \n",
       "Doc 53  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 54  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 55  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 56  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 57  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "Doc 58  0.108963  0.161876  0.185843  0.094309  0.179021  0.125477  0.090519   \n",
       "Doc 59  0.101689  0.175653  0.185424  0.136627  0.182024  0.111974  0.081340   \n",
       "\n",
       "           Doc 8     Doc 9    Doc 10  ...    Doc 50    Doc 51    Doc 52  \\\n",
       "Doc 1   0.131666  0.082319  0.103330  ...  0.158555  0.113504  0.183942   \n",
       "Doc 2   0.165781  0.222440  0.171459  ...  0.258898  0.114252  0.159238   \n",
       "Doc 3   0.155222  0.172260  0.145168  ...  0.233060  0.160828  0.175988   \n",
       "Doc 4   0.093672  0.162503  0.115923  ...  0.161168  0.089334  0.118018   \n",
       "Doc 5   0.132900  0.149990  0.128149  ...  0.193689  0.128157  0.174898   \n",
       "Doc 6   0.094784  0.082026  0.094177  ...  0.141220  0.067741  0.098841   \n",
       "Doc 7   0.135525  0.097209  0.099001  ...  0.184630  0.084729  0.173973   \n",
       "Doc 8   1.000000  0.109231  0.147724  ...  0.271639  0.089889  0.125835   \n",
       "Doc 9   0.109231  1.000000  0.118235  ...  0.165711  0.099433  0.145940   \n",
       "Doc 10  0.147724  0.118235  1.000000  ...  0.236668  0.102157  0.174861   \n",
       "Doc 11  0.097432  0.096080  0.116509  ...  0.150959  0.060228  0.150303   \n",
       "Doc 12  0.108100  0.113032  0.110699  ...  0.148205  0.075926  0.109810   \n",
       "Doc 13  0.073111  0.103513  0.097210  ...  0.131324  0.054946  0.091295   \n",
       "Doc 14  0.110996  0.121068  0.116875  ...  0.186318  0.084827  0.140457   \n",
       "Doc 15  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 16  0.166809  0.159087  0.211481  ...  0.239887  0.150346  0.207647   \n",
       "Doc 17  0.142852  0.140772  0.136021  ...  0.192152  0.114266  0.184255   \n",
       "Doc 18  0.066223  0.083643  0.097712  ...  0.131048  0.062794  0.105670   \n",
       "Doc 19  0.094137  0.117368  0.107373  ...  0.135323  0.167085  0.132014   \n",
       "Doc 20  0.105564  0.156089  0.148262  ...  0.178179  0.085107  0.149336   \n",
       "Doc 21  0.159945  0.132522  0.181965  ...  0.239480  0.102407  0.152915   \n",
       "Doc 22  0.111076  0.140537  0.145046  ...  0.160755  0.074856  0.162850   \n",
       "Doc 23  0.125713  0.138715  0.170944  ...  0.194180  0.100097  0.155831   \n",
       "Doc 24  0.098792  0.104502  0.134515  ...  0.156082  0.081920  0.143701   \n",
       "Doc 25  0.128907  0.116363  0.117308  ...  0.166920  0.116938  0.130530   \n",
       "Doc 26  0.074547  0.103582  0.070872  ...  0.118898  0.125293  0.123362   \n",
       "Doc 27  0.128998  0.094528  0.136686  ...  0.180698  0.100513  0.163864   \n",
       "Doc 28  0.116039  0.105223  0.113288  ...  0.166193  0.067261  0.133989   \n",
       "Doc 29  0.100485  0.096546  0.107054  ...  0.164877  0.082074  0.125202   \n",
       "Doc 30  0.160880  0.129961  0.206259  ...  0.222120  0.103976  0.183047   \n",
       "Doc 31  0.104529  0.120069  0.122733  ...  0.163592  0.097858  0.143472   \n",
       "Doc 32  0.141983  0.129830  0.235001  ...  0.290518  0.100291  0.182947   \n",
       "Doc 33  0.168514  0.131007  0.207040  ...  0.204532  0.104100  0.172836   \n",
       "Doc 34  0.133625  0.157121  0.216711  ...  0.296036  0.113282  0.213479   \n",
       "Doc 35  0.131145  0.110865  0.115268  ...  0.174731  0.086845  0.133350   \n",
       "Doc 36  0.111982  0.129016  0.197003  ...  0.214439  0.095076  0.164451   \n",
       "Doc 37  0.138726  0.138802  0.122547  ...  0.186266  0.135514  0.139725   \n",
       "Doc 38  0.068699  0.063132  0.069004  ...  0.101152  0.062345  0.087919   \n",
       "Doc 39  0.090382  0.082448  0.095109  ...  0.158367  0.067036  0.112057   \n",
       "Doc 40  0.112792  0.117850  0.175680  ...  0.202966  0.118186  0.162316   \n",
       "Doc 41  0.140819  0.127682  0.148133  ...  0.176488  0.108562  0.173651   \n",
       "Doc 42  0.029710  0.058840  0.121969  ...  0.166598  0.051721  0.075256   \n",
       "Doc 43  0.092156  0.091469  0.085491  ...  0.129270  0.097456  0.261086   \n",
       "Doc 44  0.108924  0.053782  0.110771  ...  0.172829  0.054417  0.056506   \n",
       "Doc 45  0.145881  0.130406  0.108719  ...  0.200937  0.166008  0.150386   \n",
       "Doc 46  0.146149  0.119740  0.150022  ...  0.187400  0.088309  0.138912   \n",
       "Doc 47  0.081322  0.098389  0.129807  ...  0.140852  0.076504  0.136313   \n",
       "Doc 48  0.086895  0.105628  0.097056  ...  0.133022  0.076530  0.111074   \n",
       "Doc 49  0.103991  0.137495  0.218288  ...  0.325398  0.072165  0.183243   \n",
       "Doc 50  0.271639  0.165711  0.236668  ...  1.000000  0.130866  0.223219   \n",
       "Doc 51  0.089889  0.099433  0.102157  ...  0.130866  1.000000  0.121413   \n",
       "Doc 52  0.125835  0.145940  0.174861  ...  0.223219  0.121413  1.000000   \n",
       "Doc 53  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 54  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 55  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 56  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 57  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   \n",
       "Doc 58  0.152414  0.139159  0.165920  ...  0.184290  0.127521  0.162394   \n",
       "Doc 59  0.124312  0.140329  0.165466  ...  0.209632  0.127359  0.156033   \n",
       "\n",
       "        Doc 53  Doc 54  Doc 55  Doc 56  Doc 57    Doc 58    Doc 59  \n",
       "Doc 1      0.0     0.0     0.0     0.0     0.0  0.108963  0.101689  \n",
       "Doc 2      0.0     0.0     0.0     0.0     0.0  0.161876  0.175653  \n",
       "Doc 3      0.0     0.0     0.0     0.0     0.0  0.185843  0.185424  \n",
       "Doc 4      0.0     0.0     0.0     0.0     0.0  0.094309  0.136627  \n",
       "Doc 5      0.0     0.0     0.0     0.0     0.0  0.179021  0.182024  \n",
       "Doc 6      0.0     0.0     0.0     0.0     0.0  0.125477  0.111974  \n",
       "Doc 7      0.0     0.0     0.0     0.0     0.0  0.090519  0.081340  \n",
       "Doc 8      0.0     0.0     0.0     0.0     0.0  0.152414  0.124312  \n",
       "Doc 9      0.0     0.0     0.0     0.0     0.0  0.139159  0.140329  \n",
       "Doc 10     0.0     0.0     0.0     0.0     0.0  0.165920  0.165466  \n",
       "Doc 11     0.0     0.0     0.0     0.0     0.0  0.174335  0.112852  \n",
       "Doc 12     0.0     0.0     0.0     0.0     0.0  0.148582  0.120566  \n",
       "Doc 13     0.0     0.0     0.0     0.0     0.0  0.135736  0.106503  \n",
       "Doc 14     0.0     0.0     0.0     0.0     0.0  0.138115  0.155765  \n",
       "Doc 15     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 16     0.0     0.0     0.0     0.0     0.0  0.209576  0.256975  \n",
       "Doc 17     0.0     0.0     0.0     0.0     0.0  0.160721  0.125082  \n",
       "Doc 18     0.0     0.0     0.0     0.0     0.0  0.077557  0.125615  \n",
       "Doc 19     0.0     0.0     0.0     0.0     0.0  0.175687  0.121154  \n",
       "Doc 20     0.0     0.0     0.0     0.0     0.0  0.177183  0.181017  \n",
       "Doc 21     0.0     0.0     0.0     0.0     0.0  0.142773  0.170131  \n",
       "Doc 22     0.0     0.0     0.0     0.0     0.0  0.136043  0.119413  \n",
       "Doc 23     0.0     0.0     0.0     0.0     0.0  0.204906  0.192122  \n",
       "Doc 24     0.0     0.0     0.0     0.0     0.0  0.186991  0.140449  \n",
       "Doc 25     0.0     0.0     0.0     0.0     0.0  0.176602  0.127690  \n",
       "Doc 26     0.0     0.0     0.0     0.0     0.0  0.100963  0.090770  \n",
       "Doc 27     0.0     0.0     0.0     0.0     0.0  0.146929  0.182905  \n",
       "Doc 28     0.0     0.0     0.0     0.0     0.0  0.150759  0.136666  \n",
       "Doc 29     0.0     0.0     0.0     0.0     0.0  0.138711  0.129555  \n",
       "Doc 30     0.0     0.0     0.0     0.0     0.0  0.209475  0.161882  \n",
       "Doc 31     0.0     0.0     0.0     0.0     0.0  0.151850  0.102236  \n",
       "Doc 32     0.0     0.0     0.0     0.0     0.0  0.170552  0.152193  \n",
       "Doc 33     0.0     0.0     0.0     0.0     0.0  0.160454  0.159194  \n",
       "Doc 34     0.0     0.0     0.0     0.0     0.0  0.158022  0.182414  \n",
       "Doc 35     0.0     0.0     0.0     0.0     0.0  0.140525  0.123981  \n",
       "Doc 36     0.0     0.0     0.0     0.0     0.0  0.172902  0.206983  \n",
       "Doc 37     0.0     0.0     0.0     0.0     0.0  0.170663  0.127123  \n",
       "Doc 38     0.0     0.0     0.0     0.0     0.0  0.081703  0.071785  \n",
       "Doc 39     0.0     0.0     0.0     0.0     0.0  0.131828  0.153358  \n",
       "Doc 40     0.0     0.0     0.0     0.0     0.0  0.137983  0.211101  \n",
       "Doc 41     0.0     0.0     0.0     0.0     0.0  0.261871  0.116705  \n",
       "Doc 42     0.0     0.0     0.0     0.0     0.0  0.046876  0.055075  \n",
       "Doc 43     0.0     0.0     0.0     0.0     0.0  0.091827  0.098842  \n",
       "Doc 44     0.0     0.0     0.0     0.0     0.0  0.088176  0.076290  \n",
       "Doc 45     0.0     0.0     0.0     0.0     0.0  0.147125  0.144709  \n",
       "Doc 46     0.0     0.0     0.0     0.0     0.0  0.201635  0.133507  \n",
       "Doc 47     0.0     0.0     0.0     0.0     0.0  0.099747  0.103475  \n",
       "Doc 48     0.0     0.0     0.0     0.0     0.0  0.083653  0.119951  \n",
       "Doc 49     0.0     0.0     0.0     0.0     0.0  0.129795  0.153425  \n",
       "Doc 50     0.0     0.0     0.0     0.0     0.0  0.184290  0.209632  \n",
       "Doc 51     0.0     0.0     0.0     0.0     0.0  0.127521  0.127359  \n",
       "Doc 52     0.0     0.0     0.0     0.0     0.0  0.162394  0.156033  \n",
       "Doc 53     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 54     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 55     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 56     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 57     0.0     0.0     0.0     0.0     0.0  0.000000  0.000000  \n",
       "Doc 58     0.0     0.0     0.0     0.0     0.0  1.000000  0.214111  \n",
       "Doc 59     0.0     0.0     0.0     0.0     0.0  0.214111  1.000000  \n",
       "\n",
       "[59 rows x 59 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "\n",
    "def similarity(tfidf_matrix):\n",
    "    # calcoliamo la cosine similarity tra i documenti\n",
    "    cos_sim = cosine_similarity(tfidf_matrix)\n",
    "\n",
    "    # creiamo una tabella con le cosine similarity per ogni coppia di documenti\n",
    "    sim_table = pd.DataFrame(cos_sim, columns=['Doc ' + str(i+1) for i in range(cos_sim.shape[0])], index=['Doc ' + str(i+1) for i in range(cos_sim.shape[0])])\n",
    "    \n",
    "    return sim_table\n",
    "\n",
    "cos_sim_table = similarity(tfidf_matrix_docs)\n",
    "cos_sim_table\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5873dd86",
   "metadata": {},
   "source": [
    "## For each document calculate the number of documents with a certain thresholds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8ef266d",
   "metadata": {},
   "source": [
    "### Sequential Algoritghm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f37df90e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 8, 7, 1, 2, 1, 1, 1, 1, 7, 2, 3, 1, 1, 0, 17, 2, 0, 0, 7, 4, 0, 4, 0, 0, 0, 3, 3, 1, 9, 2, 7, 5, 12, 0, 11, 0, 0, 2, 3, 2, 3, 2, 1, 1, 2, 0, 0, 5, 16, 0, 4, 0, 0, 0, 0, 0, 6, 5]\n"
     ]
    }
   ],
   "source": [
    "def sequential_count_similar_documents(cos_sim_table, threshold):\n",
    "    num_similar = []\n",
    "    for i in range(cos_sim_table.shape[0]):\n",
    "        num = 0\n",
    "        for j in range(cos_sim_table.shape[1]):\n",
    "            if i != j and cos_sim_table.iloc[i,j] > threshold:\n",
    "                num += 1\n",
    "        num_similar.append(num)\n",
    "    return num_similar, threshold\n",
    "\n",
    "num_similar, threshold = sequential_count_similar_documents(cos_sim_table, 0.2)\n",
    "print(num_similar)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a22b7bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Documents</th>\n",
       "      <th>Number of similar documents</th>\n",
       "      <th>Threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Doc 1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Doc 2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Doc 3</td>\n",
       "      <td>7</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Doc 4</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Doc 5</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Doc 6</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Doc 7</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Doc 8</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Doc 9</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Doc 10</td>\n",
       "      <td>7</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Doc 11</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Doc 12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Doc 13</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Doc 14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Doc 15</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Doc 16</td>\n",
       "      <td>17</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Doc 17</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Doc 18</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Doc 19</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Doc 20</td>\n",
       "      <td>7</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Doc 21</td>\n",
       "      <td>4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Doc 22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Doc 23</td>\n",
       "      <td>4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Doc 24</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Doc 25</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Doc 26</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Doc 27</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Doc 28</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Doc 29</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Doc 30</td>\n",
       "      <td>9</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Doc 31</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Doc 32</td>\n",
       "      <td>7</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Doc 33</td>\n",
       "      <td>5</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Doc 34</td>\n",
       "      <td>12</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Doc 35</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Doc 36</td>\n",
       "      <td>11</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Doc 37</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Doc 38</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Doc 39</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Doc 40</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Doc 41</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Doc 42</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Doc 43</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Doc 44</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Doc 45</td>\n",
       "      <td>1</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Doc 46</td>\n",
       "      <td>2</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Doc 47</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Doc 48</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Doc 49</td>\n",
       "      <td>5</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Doc 50</td>\n",
       "      <td>16</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Doc 51</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Doc 52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Doc 53</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Doc 54</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Doc 55</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Doc 56</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Doc 57</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Doc 58</td>\n",
       "      <td>6</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Doc 59</td>\n",
       "      <td>5</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Documents  Number of similar documents  Threshold\n",
       "0      Doc 1                            1        0.2\n",
       "1      Doc 2                            8        0.2\n",
       "2      Doc 3                            7        0.2\n",
       "3      Doc 4                            1        0.2\n",
       "4      Doc 5                            2        0.2\n",
       "5      Doc 6                            1        0.2\n",
       "6      Doc 7                            1        0.2\n",
       "7      Doc 8                            1        0.2\n",
       "8      Doc 9                            1        0.2\n",
       "9     Doc 10                            7        0.2\n",
       "10    Doc 11                            2        0.2\n",
       "11    Doc 12                            3        0.2\n",
       "12    Doc 13                            1        0.2\n",
       "13    Doc 14                            1        0.2\n",
       "14    Doc 15                            0        0.2\n",
       "15    Doc 16                           17        0.2\n",
       "16    Doc 17                            2        0.2\n",
       "17    Doc 18                            0        0.2\n",
       "18    Doc 19                            0        0.2\n",
       "19    Doc 20                            7        0.2\n",
       "20    Doc 21                            4        0.2\n",
       "21    Doc 22                            0        0.2\n",
       "22    Doc 23                            4        0.2\n",
       "23    Doc 24                            0        0.2\n",
       "24    Doc 25                            0        0.2\n",
       "25    Doc 26                            0        0.2\n",
       "26    Doc 27                            3        0.2\n",
       "27    Doc 28                            3        0.2\n",
       "28    Doc 29                            1        0.2\n",
       "29    Doc 30                            9        0.2\n",
       "30    Doc 31                            2        0.2\n",
       "31    Doc 32                            7        0.2\n",
       "32    Doc 33                            5        0.2\n",
       "33    Doc 34                           12        0.2\n",
       "34    Doc 35                            0        0.2\n",
       "35    Doc 36                           11        0.2\n",
       "36    Doc 37                            0        0.2\n",
       "37    Doc 38                            0        0.2\n",
       "38    Doc 39                            2        0.2\n",
       "39    Doc 40                            3        0.2\n",
       "40    Doc 41                            2        0.2\n",
       "41    Doc 42                            3        0.2\n",
       "42    Doc 43                            2        0.2\n",
       "43    Doc 44                            1        0.2\n",
       "44    Doc 45                            1        0.2\n",
       "45    Doc 46                            2        0.2\n",
       "46    Doc 47                            0        0.2\n",
       "47    Doc 48                            0        0.2\n",
       "48    Doc 49                            5        0.2\n",
       "49    Doc 50                           16        0.2\n",
       "50    Doc 51                            0        0.2\n",
       "51    Doc 52                            4        0.2\n",
       "52    Doc 53                            0        0.2\n",
       "53    Doc 54                            0        0.2\n",
       "54    Doc 55                            0        0.2\n",
       "55    Doc 56                            0        0.2\n",
       "56    Doc 57                            0        0.2\n",
       "57    Doc 58                            6        0.2\n",
       "58    Doc 59                            5        0.2"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def create_similar_table(num_similar, threshold):\n",
    "    doc_names = [f\"Doc {i+1}\" for i in range(len(num_similar))]\n",
    "    similar_table = pd.DataFrame({\"Documents\": doc_names, \"Number of similar documents\": num_similar, \"Threshold\":threshold })\n",
    "    return similar_table\n",
    "\n",
    "\n",
    "similar_table = create_similar_table(num_similar, threshold)\n",
    "similar_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2900c4d6",
   "metadata": {},
   "source": [
    "# Prove Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "280d84a8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cluster 1': ['Doc 43', 'Doc 1'],\n",
       " 'Cluster 2': ['Doc 3',\n",
       "  'Doc 4',\n",
       "  'Doc 9',\n",
       "  'Doc 30',\n",
       "  'Doc 33',\n",
       "  'Doc 34',\n",
       "  'Doc 36',\n",
       "  'Doc 50',\n",
       "  'Doc 2'],\n",
       " 'Cluster 3': ['Doc 16', 'Doc 5'],\n",
       " 'Cluster 4': ['Doc 6'],\n",
       " 'Cluster 5': ['Doc 17', 'Doc 7'],\n",
       " 'Cluster 6': ['Doc 8'],\n",
       " 'Cluster 7': ['Doc 32', 'Doc 49', 'Doc 10'],\n",
       " 'Cluster 8': ['Doc 28', 'Doc 31', 'Doc 11'],\n",
       " 'Cluster 9': ['Doc 20', 'Doc 12'],\n",
       " 'Cluster 10': ['Doc 13'],\n",
       " 'Cluster 11': ['Doc 14'],\n",
       " 'Cluster 12': ['Doc 15'],\n",
       " 'Cluster 13': ['Doc 18'],\n",
       " 'Cluster 14': ['Doc 19'],\n",
       " 'Cluster 15': ['Doc 21'],\n",
       " 'Cluster 16': ['Doc 22'],\n",
       " 'Cluster 17': ['Doc 39', 'Doc 58', 'Doc 23'],\n",
       " 'Cluster 18': ['Doc 24'],\n",
       " 'Cluster 19': ['Doc 25'],\n",
       " 'Cluster 20': ['Doc 26'],\n",
       " 'Cluster 21': ['Doc 27'],\n",
       " 'Cluster 22': ['Doc 29'],\n",
       " 'Cluster 23': ['Doc 35'],\n",
       " 'Cluster 24': ['Doc 37'],\n",
       " 'Cluster 25': ['Doc 38'],\n",
       " 'Cluster 26': ['Doc 59', 'Doc 40'],\n",
       " 'Cluster 27': ['Doc 41'],\n",
       " 'Cluster 28': ['Doc 42'],\n",
       " 'Cluster 29': ['Doc 44'],\n",
       " 'Cluster 30': ['Doc 45'],\n",
       " 'Cluster 31': ['Doc 46'],\n",
       " 'Cluster 32': ['Doc 47'],\n",
       " 'Cluster 33': ['Doc 48'],\n",
       " 'Cluster 34': ['Doc 51'],\n",
       " 'Cluster 35': ['Doc 52'],\n",
       " 'Cluster 36': ['Doc 53'],\n",
       " 'Cluster 37': ['Doc 54'],\n",
       " 'Cluster 38': ['Doc 55'],\n",
       " 'Cluster 39': ['Doc 56'],\n",
       " 'Cluster 40': ['Doc 57']}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def divide_documents(cos_sim_table, threshold):\n",
    "    # Creiamo un dizionario per i cluster\n",
    "    clusters = {}\n",
    "    # Creiamo un set per tenere traccia dei documenti già assegnati a un cluster\n",
    "    assigned_docs = set()\n",
    "    # Iteriamo su ogni riga e colonna della tabella di similarità\n",
    "    for i, row in cos_sim_table.iterrows():\n",
    "        if i not in assigned_docs:  # Se il documento non è già stato assegnato a un cluster\n",
    "            cluster = []  # Creiamo un nuovo cluster\n",
    "            for j, sim in row.iteritems():\n",
    "                if j != i and j not in assigned_docs and sim >= threshold:  # Se il documento non è già stato assegnato a un cluster, non è se stesso e ha una similarità maggiore o uguale alla soglia\n",
    "                    cluster.append(j)  # Aggiungiamo il documento al cluster\n",
    "                    assigned_docs.add(j)  # Segniamo il documento come assegnato\n",
    "            cluster.append(i)  # Aggiungiamo il documento corrente al cluster\n",
    "            assigned_docs.add(i)  # Segniamo il documento corrente come assegnato\n",
    "            clusters[f'Cluster {len(clusters)+1}'] = cluster  # Aggiungiamo il cluster al dizionario di cluster\n",
    "    return clusters\n",
    "\n",
    "\n",
    "clusters = divide_documents(cos_sim_table, 0.2)\n",
    "clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29991013",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d0fa1e27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "def find_similar_documents(cos_sim_table, doc_name, similarity_threshold):\n",
    "    # trova il vettore di similarità del documento specificato\n",
    "    doc_sim_vector = cos_sim_table[doc_name]\n",
    "\n",
    "    # trova il numero di documenti che hanno una similarità superiore alla soglia specificata\n",
    "    similar_docs_count = len([similarity for similarity in doc_sim_vector if similarity > similarity_threshold])\n",
    "\n",
    "    return similar_docs_count\n",
    "\n",
    "similar_docs_count = find_similar_documents(cos_sim_table, 'Doc 1', 0.2)\n",
    "print(similar_docs_count-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8272e79",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3cd7de62",
   "metadata": {},
   "source": [
    "# Parallel Alghoritm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06ff0826",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ['PYSPARK_PYTHON'] = 'C:/Users/lita4/anaconda3/python.exe'\n",
    "os.environ['PYSPARK_DRIVER_PYTHON'] = 'C:/Users/lita4/anaconda3/python.exe'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee9756d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import SparkSession\n",
    "\n",
    "spark = SparkSession.builder.appName(\"document-similarity\").getOrCreate()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b9d58896",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_name_documents = './Databases/prova/prova50.jsonl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b6a1315",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c9cc5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#conf = SparkConf().setAppName(\"MyApp\")\n",
    "#sc = SparkContext(conf=conf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "11f8274e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('ug7v899j', 'OBJECTIVE: This retrospective chart review describes the epidemiology and clinical features of 40 patients with culture-proven Mycoplasma pneumoniae infections at King Abdulaziz University Hospital, Jeddah, Saudi Arabia. METHODS: Patients with positive M. pneumoniae cultures from respiratory specimens from January 1997 through December 1998 were identified through the Microbiology records. Charts of patients were reviewed. RESULTS: 40 patients were identified, 33 (82.5%) of whom required admission. Most infections (92.5%) were community-acquired. The infection affected all age groups but was most common in infants (32.5%) and pre-school children (22.5%). It occurred year-round but was most common in the fall (35%) and spring (30%). More than three-quarters of patients (77.5%) had comorbidities. Twenty-four isolates (60%) were associated with pneumonia, 14 (35%) with upper respiratory tract infections, and 2 (5%) with bronchiolitis. Cough (82.5%), fever (75%), and malaise (58.8%) were the most common symptoms, and crepitations (60%), and wheezes (40%) were the most common signs. Most patients with pneumonia had crepitations (79.2%) but only 25% had bronchial breathing. Immunocompromised patients were more likely than non-immunocompromised patients to present with pneumonia (8/9 versus 16/31, P = 0.05). Of the 24 patients with pneumonia, 14 (58.3%) had uneventful recovery, 4 (16.7%) recovered following some complications, 3 (12.5%) died because of M pneumoniae infection, and 3 (12.5%) died due to underlying comorbidities. The 3 patients who died of M pneumoniae pneumonia had other comorbidities. CONCLUSION: our results were similar to published data except for the finding that infections were more common in infants and preschool children and that the mortality rate of pneumonia in patients with comorbidities was high.'), ('02tnwd4m', 'Inflammatory diseases of the respiratory tract are commonly associated with elevated production of nitric oxide (NO•) and increased indices of NO• -dependent oxidative stress. Although NO• is known to have anti-microbial, anti-inflammatory and anti-oxidant properties, various lines of evidence support the contribution of NO• to lung injury in several disease models. On the basis of biochemical evidence, it is often presumed that such NO• -dependent oxidations are due to the formation of the oxidant peroxynitrite, although alternative mechanisms involving the phagocyte-derived heme proteins myeloperoxidase and eosinophil peroxidase might be operative during conditions of inflammation. Because of the overwhelming literature on NO• generation and activities in the respiratory tract, it would be beyond the scope of this commentary to review this area comprehensively. Instead, it focuses on recent evidence and concepts of the presumed contribution of NO• to inflammatory diseases of the lung.'), ('ejv2xln0', 'Surfactant protein-D (SP-D) participates in the innate response to inhaled microorganisms and organic antigens, and contributes to immune and inflammatory regulation within the lung. SP-D is synthesized and secreted by alveolar and bronchiolar epithelial cells, but is also expressed by epithelial cells lining various exocrine ducts and the mucosa of the gastrointestinal and genitourinary tracts. SP-D, a collagenous calcium-dependent lectin (or collectin), binds to surface glycoconjugates expressed by a wide variety of microorganisms, and to oligosaccharides associated with the surface of various complex organic antigens. SP-D also specifically interacts with glycoconjugates and other molecules expressed on the surface of macrophages, neutrophils, and lymphocytes. In addition, SP-D binds to specific surfactant-associated lipids and can influence the organization of lipid mixtures containing phosphatidylinositol in vitro. Consistent with these diverse in vitro activities is the observation that SP-D-deficient transgenic mice show abnormal accumulations of surfactant lipids, and respond abnormally to challenge with respiratory viruses and bacterial lipopolysaccharides. The phenotype of macrophages isolated from the lungs of SP-D-deficient mice is altered, and there is circumstantial evidence that abnormal oxidant metabolism and/or increased metalloproteinase expression contributes to the development of emphysema. The expression of SP-D is increased in response to many forms of lung injury, and deficient accumulation of appropriately oligomerized SP-D might contribute to the pathogenesis of a variety of human lung diseases.'), ('2b73a28n', 'Endothelin-1 (ET-1) is a 21 amino acid peptide with diverse biological activity that has been implicated in numerous diseases. ET-1 is a potent mitogen regulator of smooth muscle tone, and inflammatory mediator that may play a key role in diseases of the airways, pulmonary circulation, and inflammatory lung diseases, both acute and chronic. This review will focus on the biology of ET-1 and its role in lung disease.'), ('9785vg6d', 'Respiratory syncytial virus (RSV) and pneumonia virus of mice (PVM) are viruses of the family Paramyxoviridae, subfamily pneumovirus, which cause clinically important respiratory infections in humans and rodents, respectively. The respiratory epithelial target cells respond to viral infection with specific alterations in gene expression, including production of chemoattractant cytokines, adhesion molecules, elements that are related to the apoptosis response, and others that remain incompletely understood. Here we review our current understanding of these mucosal responses and discuss several genomic approaches, including differential display reverse transcription-polymerase chain reaction (PCR) and gene array strategies, that will permit us to unravel the nature of these responses in a more complete and systematic manner.'), ('zjufx4fo', 'Nidovirus subgenomic mRNAs contain a leader sequence derived from the 5′ end of the genome fused to different sequences (‘bodies’) derived from the 3′ end. Their generation involves a unique mechanism of discontinuous subgenomic RNA synthesis that resembles copy-choice RNA recombination. During this process, the nascent RNA strand is transferred from one site in the template to another, during either plus or minus strand synthesis, to yield subgenomic RNA molecules. Central to this process are transcription-regulating sequences (TRSs), which are present at both template sites and ensure the fidelity of strand transfer. Here we present results of a comprehensive co-variation mutagenesis study of equine arteritis virus TRSs, demonstrating that discontinuous RNA synthesis depends not only on base pairing between sense leader TRS and antisense body TRS, but also on the primary sequence of the body TRS. While the leader TRS merely plays a targeting role for strand transfer, the body TRS fulfils multiple functions. The sequences of mRNA leader–body junctions of TRS mutants strongly suggested that the discontinuous step occurs during minus strand synthesis.'), ('5yhe786e', 'Recent evidence suggests that critically ill patients are able to tolerate lower levels of haemoglobin than was previously believed. It is our goal to show that transfusing to a level of 100 g/l does not improve mortality and other clinically important outcomes in a critical care setting. Although many questions remain, many laboratory and clinical studies, including a recent randomized controlled trial (RCT), have established that transfusing to normal haemoglobin concentrations does not improve organ failure and mortality in the critically ill patient. In addition, a restrictive transfusion strategy will reduce exposure to allogeneic transfusions, result in more efficient use of red blood cells (RBCs), save blood overall, and decrease health care costs.'), ('8zchiykl', \"The 21st International Symposium on Intensive Care and Emergency Medicine was dominated by the results of recent clinical trials in sepsis and acute respiratory distress syndrome (ARDS). The promise of extracorporeal liver replacement therapy and noninvasive ventilation were other areas of interest. Ethical issues also received attention. Overall, the 'state of the art' lectures, pro/con debates, seminars and tutorials were of a high standard. The meeting was marked by a sense of renewed enthusiasm that positive progress is occurring in intensive care medicine.\"), ('8qnrcgnk', 'Heme oxygenase-1 (HO-1), an inducible stress protein, confers cytoprotection against oxidative stress in vitro and in vivo. In addition to its physiological role in heme degradation, HO-1 may influence a number of cellular processes, including growth, inflammation, and apoptosis. By virtue of anti-inflammatory effects, HO-1 limits tissue damage in response to proinflammatory stimuli and prevents allograft rejection after transplantation. The transcriptional upregulation of HO-1 responds to many agents, such as hypoxia, bacterial lipopolysaccharide, and reactive oxygen/nitrogen species. HO-1 and its constitutively expressed isozyme, heme oxygenase-2, catalyze the rate-limiting step in the conversion of heme to its metabolites, bilirubin IXα, ferrous iron, and carbon monoxide (CO). The mechanisms by which HO-1 provides protection most likely involve its enzymatic reaction products. Remarkably, administration of CO at low concentrations can substitute for HO-1 with respect to anti-inflammatory and anti-apoptotic effects, suggesting a role for CO as a key mediator of HO-1 function. Chronic, low-level, exogenous exposure to CO from cigarette smoking contributes to the importance of CO in pulmonary medicine. The implications of the HO-1/CO system in pulmonary diseases will be discussed in this review, with an emphasis on inflammatory states.'), ('jg13scgo', 'This report describes the design and implementation of the Real-time Outbreak and Disease Surveillance (RODS) system, a computer-based public health surveillance system for early detection of disease outbreaks. Hospitals send RODS data from clinical encounters over virtual private networks and leased lines using the Health Level 7 (HL7) message protocol. The data are sent in real time. RODS automatically classifies the registration chief complaint from the visit into one of seven syndrome categories using Bayesian classifiers. It stores the data in a relational database, aggregates the data for analysis using data warehousing techniques, applies univariate and multivariate statistical detection algorithms to the data, and alerts users of when the algorithms identify anomalous patterns in the syndrome counts. RODS also has a Web-based user interface that supports temporal and spatial analyses. RODS processes sales of over-the-counter health care products in a similar manner but receives such data in batch mode on a daily basis. RODS was used during the 2002 Winter Olympics and currently operates in two states—Pennsylvania and Utah. It has been and continues to be a resource for implementing, evaluating, and applying new methods of public health surveillance.'), ('5tkvsudh', 'Regulation of ornithine decarboxylase in vertebrates involves a negative feedback mechanism requiring the protein antizyme. Here we show that a similar mechanism exists in the fission yeast Schizosaccharomyces pombe. The expression of mammalian antizyme genes requires a specific +1 translational frameshift. The efficiency of the frameshift event reflects cellular polyamine levels creating the autoregulatory feedback loop. As shown here, the yeast antizyme gene and several newly identified antizyme genes from different nematodes also require a ribosomal frameshift event for their expression. Twelve nucleotides around the frameshift site are identical between S.pombe and the mammalian counterparts. The core element for this frameshifting is likely to have been present in the last common ancestor of yeast, nematodes and mammals.'), ('6lvn10f4', 'Heterogeneous nuclear ribonucleoprotein (hnRNP A1) is involved in pre-mRNA splicing in the nucleus and translational regulation in the cytoplasm. In the present study, we demonstrate that hnRNP A1 also participates in the transcription and replication of a cytoplasmic RNA virus, mouse hepatitis virus (MHV). Overexpression of hnRNP A1 accelerated the kinetics of viral RNA synthesis, whereas the expression in the cytoplasm of a dominant-negative hnRNP A1 mutant that lacks the nuclear transport domain significantly delayed it. The hnRNP A1 mutant caused a global inhibition of viral mRNA transcription and genomic replication, and also a preferential inhibition of the replication of defective-interfering RNAs. Similar to the wild-type hnRNP A1, the hnRNP A1 mutant complexed with an MHV polymerase gene product, the nucleocapsid protein and the viral RNA. However, in contrast to the wild-type hnRNP A1, the mutant protein failed to bind a 250 kDa cellular protein, suggesting that the recruitment of cellular proteins by hnRNP A1 is important for MHV RNA synthesis. Our findings establish the importance of cellular factors in viral RNA-dependent RNA synthesis.'), ('tvxpckxo', 'The UBA domain is a conserved sequence motif among polyubiquitin binding proteins. For the first time, we demonstrate a systematic, high throughput approach to identification of UBA domain-interacting proteins from a proteome-wide perspective. Using the rabbit reticulocyte lysate in vitro expression cloning system, we have successfully identified eleven proteins that interact with p62’s UBA domain, and the majority of the eleven proteins are associated with neurodegenerative disorders, such as Alzheimer’s disease. Therefore, p62 may play a novel regulatory role through its UBA domain. Our approach provides an easy route to the characterization of UBA domain interacting proteins and its application will unfold the important roles that the UBA domain plays.'), ('mcuixluu', 'We examined the role of the microtubule cytoskeleton during vaccinia virus infection. We found that newly assembled virus particles accumulate in the vicinity of the microtubule-organizing centre in a microtubule- and dynein–dynactin complex-dependent fashion. Microtubules are required for efficient intracellular mature virus (IMV) formation and are essential for intracellular enveloped virus (IEV) assembly. As infection proceeds, the microtubule cytoskeleton becomes dramatically reorganized in a fashion reminiscent of overexpression of microtubule-associated proteins (MAPs). Consistent with this, we report that the vaccinia proteins A10L and L4R have MAP-like properties and mediate direct binding of viral cores to microtubules in vitro. In addition, vaccinia infection also results in severe reduction of proteins at the centrosome and loss of centrosomal microtubule nucleation efficiency. This represents the first example of viral-induced disruption of centrosome function. Further studies with vaccinia will provide insights into the role of microtubules during viral pathogenesis and regulation of centrosome function.'), ('6iu1dtyl', ''), ('t35n7bk9', 'There are hundreds of viruses that infect different human organs and cause diseases. Some fatal emerging viral infections have become serious public health issues worldwide. Early diagnosis and subsequent treatment are therefore essential for fighting viral infections. Current diagnostic techniques frequently employ polymerase chain reaction (PCR)-based methods to quickly detect the pathogenic viruses and establish the etiology of the disease or illness. However, the fast PCR method suffers from many drawbacks such as a high false-positive rate and the ability to detect only one or a few gene targets at a time. Microarray technology solves the problems of the PCR limitations and can be effectively applied to all fields of molecular medicine. Recently, a report in Retrovirology described a multi-virus DNA array that contains more than 250 open reading frames from eight human viruses including human immunodeficiency virus type 1. This array can be used to detect multiple viral co-infections in cells and in vivo. Another benefit of this kind of multi-virus array is in studying promoter activity and viral gene expression and correlating such readouts with the progression of disease and reactivation of latent infections. Thus, the virus DNA-chip development reported in Retrovirology is an important advance in diagnostic application which could be a potent clinical tool for characterizing viral co-infections in AIDS as well as other patients.'), ('eiqypt0m', 'INTRODUCTION: The pathogenicity of late respiratory infections with herpes simplex virus type 1 (HSV-1) in the critically ill is unclear. METHODS: In four critically ill patients with persistent pulmonary infiltrates of unknown origin and isolation of HSV-1 from tracheal aspirate or bronchoalveolar lavage fluid, at 7 (1–11) days after start of mechanical ventilatory support, a pulmonary leak index (PLI) for (67)Gallium ((67)Ga)-transferrin (upper limit of normal 14.1 × 10(-3)/min) was measured. RESULTS: The PLI ranged between 7.5 and 14.0 × 10(-3)/min in the study patients. Two patients received a course of acyclovir and all survived. CONCLUSIONS: The normal capillary permeability observed in the lungs argues against pathogenicity of HSV-1 in the critically ill, and favors that isolation of the virus reflects reactivation in the course of serious illness and immunodepresssion, rather than primary or superimposed infection in the lungs.'), ('sgmk96vr', 'BACKGROUND: Previous smallpox ring vaccination models based on contact tracing over a network suggest that ring vaccination would be effective, but have not explicitly included response logistics and limited numbers of vaccinators. METHODS: We developed a continuous-time stochastic simulation of smallpox transmission, including network structure, post-exposure vaccination, vaccination of contacts of contacts, limited response capacity, heterogeneity in symptoms and infectiousness, vaccination prior to the discontinuation of routine vaccination, more rapid diagnosis due to public awareness, surveillance of asymptomatic contacts, and isolation of cases. RESULTS: We found that even in cases of very rapidly spreading smallpox, ring vaccination (when coupled with surveillance) is sufficient in most cases to eliminate smallpox quickly, assuming that 95% of household contacts are traced, 80% of workplace or social contacts are traced, and no casual contacts are traced, and that in most cases the ability to trace 1–5 individuals per day per index case is sufficient. If smallpox is assumed to be transmitted very quickly to contacts, it may at times escape containment by ring vaccination, but could be controlled in these circumstances by mass vaccination. CONCLUSIONS: Small introductions of smallpox are likely to be easily contained by ring vaccination, provided contact tracing is feasible. Uncertainties in the nature of bioterrorist smallpox (infectiousness, vaccine efficacy) support continued planning for ring vaccination as well as mass vaccination. If initiated, ring vaccination should be conducted without delays in vaccination, should include contacts of contacts (whenever there is sufficient capacity) and should be accompanied by increased public awareness and surveillance.'), ('di0fcy0j', \"BACKGROUND: Oxygen toxicity is a major cause of lung injury. The base excision repair pathway is one of the most important cellular protection mechanisms that responds to oxidative DNA damage. Lesion-specific DNA repair enzymes include hOgg1, hMYH, hNTH and hMTH. METHODS: The above lesion-specific DNA repair enzymes were expressed in human alveolar epithelial cells (A549) using the pSF91.1 retroviral vector. Cells were exposed to a 95% oxygen environment, ionizing radiation (IR), or H(2)O(2). Cell growth analysis was performed under non-toxic conditions. Western blot analysis was performed to verify over-expression and assess endogenous expression under toxic and non-toxic conditions. Statistical analysis was performed using the paired Student's t test with significance being accepted for p < 0.05. RESULTS: Cell killing assays demonstrated cells over-expressing hMYH had improved survival to both increased oxygen and IR. Cell growth analysis of A549 cells under non-toxic conditions revealed cells over-expressing hMYH also grow at a slower rate. Western blot analysis demonstrated over-expression of each individual gene and did not result in altered endogenous expression of the others. However, it was observed that O(2 )toxicity did lead to a reduced endogenous expression of hNTH in A549 cells. CONCLUSION: Increased expression of the DNA glycosylase repair enzyme hMYH in A549 cells exposed to O(2 )and IR leads to improvements in cell survival. DNA repair through the base excision repair pathway may provide an alternative way to offset the damaging effects of O(2 )and its metabolites.\"), ('4k8f7ou1', 'BACKGROUND: AlkB-like proteins are members of the 2-oxoglutarate- and Fe(II)-dependent oxygenase superfamily. In Escherichia coli the protein protects RNA and DNA against damage from methylating agents. 1-methyladenine and 3-methylcytosine are repaired by oxidative demethylation and direct reversal of the methylated base back to its unmethylated form. Genes for AlkB homologues are widespread in nature, and Eukaryotes often have several genes coding for AlkB-like proteins. Similar domains have also been observed in certain plant viruses. The function of the viral domain is unknown, but it has been suggested that it may be involved in protecting the virus against the post-transcriptional gene silencing (PTGS) system found in plants. We wanted to do a phylogenomic mapping of viral AlkB-like domains as a basis for analysing functional aspects of these domains, because this could have some relevance for understanding possible alternative roles of AlkB homologues e.g. in Eukaryotes. RESULTS: Profile-based searches of protein sequence libraries showed that AlkB-like domains are found in at least 22 different single-stranded RNA positive-strand plant viruses, but mainly in a subgroup of the Flexiviridae family. Sequence analysis indicated that the AlkB domains probably are functionally conserved, and that they most likely have been integrated relatively recently into several viral genomes at geographically distinct locations. This pattern seems to be more consistent with increased environmental pressure, e.g. from methylating pesticides, than with interaction with the PTGS system. CONCLUSIONS: The AlkB domain found in viral genomes is most likely a conventional DNA/RNA repair domain that protects the viral RNA genome against methylating compounds from the environment.'), ('wnnsmx60', \"In the 1980's and 1990's HIV/AIDS was the emerging infectious disease. In 2003–2004 we saw the emergence of SARS, Avian influenza and Anthrax in a man made form used for bioterrorism. Emergency powers legislation in Australia is a patchwork of Commonwealth quarantine laws and State and Territory based emergency powers in public health legislation. It is time for a review of such legislation and time for consideration of the efficacy of such legislation from a country wide perspective in an age when we have to consider the possibility of mass outbreaks of communicable diseases which ignore jurisdictional boundaries.\"), ('gdsfkw1b', 'Lactococcus lactis, the model lactic acid bacterium (LAB), is a food grade and well-characterized Gram positive bacterium. It is a good candidate for heterologous protein delivery in foodstuff or in the digestive tract. L. lactis can also be used as a protein producer in fermentor. Many heterologous proteins have already been produced in L. lactis but only few reports allow comparing production yields for a given protein either produced intracellularly or secreted in the medium. Here, we review several works evaluating the influence of the localization on the production yields of several heterologous proteins produced in L. lactis. The questions of size limits, conformation, and proteolysis are addressed and discussed with regard to protein yields. These data show that i) secretion is preferable to cytoplasmic production; ii) secretion enhancement (by signal peptide and propeptide optimization) results in increased production yield; iii) protein conformation rather than protein size can impair secretion and thus alter production yields; and iv) fusion of a stable protein can stabilize labile proteins. The role of intracellular proteolysis on heterologous cytoplasmic proteins and precursors is discussed. The new challenges now are the development of food grade systems and the identification and optimization of host factors affecting heterologous protein production not only in L. lactis, but also in other LAB species.'), ('yba7mdtb', 'Horizontal DNA transfer is an important factor of evolution and participates in biological diversity. Unfortunately, the location and length of horizontal transfers (HTs) are known for very few species. The usage of short oligonucleotides in a sequence (the so-called genomic signature) has been shown to be species-specific even in DNA fragments as short as 1 kb. The genomic signature is therefore proposed as a tool to detect HTs. Since DNA transfers originate from species with a signature different from those of the recipient species, the analysis of local variations of signature along recipient genome may allow for detecting exogenous DNA. The strategy consists in (i) scanning the genome with a sliding window, and calculating the corresponding local signature (ii) evaluating its deviation from the signature of the whole genome and (iii) looking for similar signatures in a database of genomic signatures. A total of 22 prokaryote genomes are analyzed in this way. It has been observed that atypical regions make up ∼6% of each genome on the average. Most of the claimed HTs as well as new ones are detected. The origin of putative DNA transfers is looked for among ∼12 000 species. Donor species are proposed and sometimes strongly suggested, considering similarity of signatures. Among the species studied, Bacillus subtilis, Haemophilus Influenzae and Escherichia coli are investigated by many authors and give the opportunity to perform a thorough comparison of most of the bioinformatics methods used to detect HTs.'), ('bbvxu8op', 'Although oligonucleotide probes complementary to single nucleotide substitutions are commonly used in microarray-based screens for genetic variation, little is known about the hybridization properties of probes complementary to small insertions and deletions. It is necessary to define the hybridization properties of these latter probes in order to improve the specificity and sensitivity of oligonucleotide microarray-based mutational analysis of disease-related genes. Here, we compare and contrast the hybridization properties of oligonucleotide microarrays consisting of 25mer probes complementary to all possible single nucleotide substitutions and insertions, and one and two base deletions in the 9168 bp coding region of the ATM (ataxia telangiectasia mutated) gene. Over 68 different dye-labeled single-stranded nucleic acid targets representing all ATM coding exons were applied to these microarrays. We assess hybridization specificity by comparing the relative hybridization signals from probes perfectly matched to ATM sequences to those containing mismatches. Probes complementary to two base substitutions displayed the highest average specificity followed by those complementary to single base substitutions, single base deletions and single base insertions. In all the cases, hybridization specificity was strongly influenced by sequence context and possible intra- and intermolecular probe and/or target structure. Furthermore, single nucleotide substitution probes displayed the most consistent hybridization specificity data followed by single base deletions, two base deletions and single nucleotide insertions. Overall, these studies provide valuable empirical data that can be used to more accurately model the hybridization properties of insertion and deletion probes and improve the design and interpretation of oligonucleotide microarray-based resequencing and mutational analysis.'), ('e62cfqt7', 'Using differential display RT-PCR, we identified a gene of 2750 bp from human adult testis, named H-Lse, which encoded a putative protein of 523 amino acids and molecular weight of 58 kd with structural characteristics similar to that of mouse lysosome sialic-acid-specific 9-O-acetylesterase. Northern blot analysis showed a widespread distribution of H-Lse in various human tissues with high expression in the testis, prostate, and colon. In situ hybridization results showed that while H-Lse was not detected in embryonic testis, positive signals were found in spermatocytes but not spermatogonia in adult testis of human. The subcellular localization of H-Lse was visualized by green fluorescent protein (GFP) fused to the amino terminus of H-Lse, showing compartmentalization of H-Lse in large dense-core vesicles, presumably lysosomes, in the cytoplasm. The developmentally regulated and spermatogenic stage-specific expression of H-Lse suggests its possible involvement in the development of the testis and/or differentiation of germ cells.'), ('4cvy9u28', 'BACKGROUND: The biological basis of pain in chronic pancreatitis is poorly understood. Mast cells have been implicated in the pathogenesis of pain in other conditions. We hypothesized that mast cells play a role in the pain of chronic pancreatitis. We examined the association of pain with mast cells in autopsy specimens of patients with painful chronic pancreatitis. We explored our hypothesis further using an experimental model of trinitrobenzene sulfonic acid (TNBS) -induced chronic pancreatitis in both wild type (WT) and mast cell deficient mice (MCDM). METHODS: Archival tissues with histological diagnoses of chronic pancreatitis were identified and clinical records reviewed for presence or absence of reported pain in humans. Mast cells were counted. The presence of pain was assessed using von Frey Filaments (VFF) to measure abdominal withdrawal responses in both WT and MCDM mice with and without chronic pancreatitis. RESULTS: Humans with painful chronic pancreatitis demonstrated a 3.5-fold increase in pancreatic mast cells as compared with those with painless chronic pancreatitis. WT mice with chronic pancreatitis were significantly more sensitive as assessed by VFF pain testing of the abdomen when compared with MCDM. CONCLUSION: Humans with painful chronic pancreatitis have an increased number of pancreatic mast cells as compared with those with painless chronic pancreatitis. MCDM are less sensitive to mechanical stimulation of the abdomen after induction of chronic pancreatitis as compared with WT. Mast cells may play an important role in the pathogenesis of pain in chronic pancreatitis.'), ('zowp10ts', 'Viral recombination can dramatically impact evolution and epidemiology. In viruses, the recombination rate depends on the frequency of genetic exchange between different viral genomes within an infected host cell and on the frequency at which such co-infections occur. While the recombination rate has been recently evaluated in experimentally co-infected cell cultures for several viruses, direct quantification at the most biologically significant level, that of a host infection, is still lacking. This study fills this gap using the cauliflower mosaic virus as a model. We distributed four neutral markers along the viral genome, and co-inoculated host plants with marker-containing and wild-type viruses. The frequency of recombinant genomes was evaluated 21 d post-inoculation. On average, over 50% of viral genomes recovered after a single host infection were recombinants, clearly indicating that recombination is very frequent in this virus. Estimates of the recombination rate show that all regions of the genome are equally affected by this process. Assuming that ten viral replication cycles occurred during our experiment—based on data on the timing of coat protein detection—the per base and replication cycle recombination rate was on the order of 2 × 10(−5) to 4 × 10(−5). This first determination of a virus recombination rate during a single multi-cellular host infection indicates that recombination is very frequent in the everyday life of this virus.'), ('5dk231qs', \"mRNA pseudoknots have a stimulatory function in programmed −1 ribosomal frameshifting (−1 PRF). Though we previously presented a model for how mRNA pseudoknots might activate the mechanism for −1 PRF, it did not address the question of the role that they may play in positioning the mRNA relative to the ribosome in this process [E. P. Plant, K. L. M. Jacobs, J. W. Harger, A. Meskauskas, J. L. Jacobs, J. L. Baxter, A. N. Petrov and J. D. Dinman (2003) RNA, 9, 168–174]. A separate ‘torsional restraint’ model suggests that mRNA pseudoknots act to increase the fraction of ribosomes directed to pause with the upstream heptameric slippery site positioned at the ribosome's A- and P-decoding sites [J. D. Dinman (1995) Yeast, 11, 1115–1127]. Here, experiments using a series of ‘pseudo-pseudoknots’ having different degrees of rotational freedom were used to test this model. The results of this study support the mechanistic hypothesis that −1 ribosomal frameshifting is enhanced by torsional resistance of the mRNA pseudoknot.\"), ('snqdma0s', 'Although efficient methods exist to assemble synthetic oligonucleotides into genes and genomes, these suffer from the presence of 1–3 random errors/kb of DNA. Here, we introduce a new method termed consensus shuffling and demonstrate its use to significantly reduce random errors in synthetic DNA. In this method, errors are revealed as mismatches by re-hybridization of the population. The DNA is fragmented, and mismatched fragments are removed upon binding to an immobilized mismatch binding protein (MutS). PCR assembly of the remaining fragments yields a new population of full-length sequences enriched for the consensus sequence of the input population. We show that two iterations of consensus shuffling improved a population of synthetic green fluorescent protein (GFPuv) clones from ∼60 to >90% fluorescent, and decreased errors 3.5- to 4.3-fold to final values of ∼1 error per 3500 bp. In addition, two iterations of consensus shuffling corrected a population of GFPuv clones where all members were non-functional, to a population where 82% of clones were fluorescent. Consensus shuffling should facilitate the rapid and accurate synthesis of long DNA sequences.'), ('1pq6dkl5', 'While it is universally accepted that intact RNA constitutes the best representation of the steady-state of transcription, there is no gold standard to define RNA quality prior to gene expression analysis. In this report, we evaluated the reliability of conventional methods for RNA quality assessment including UV spectroscopy and 28S:18S area ratios, and demonstrated their inconsistency. We then used two new freely available classifiers, the Degradometer and RIN systems, to produce user-independent RNA quality metrics, based on analysis of microcapillary electrophoresis traces. Both provided highly informative and valuable data and the results were found highly correlated, while the RIN system gave more reliable data. The relevance of the RNA quality metrics for assessment of gene expression differences was tested by Q-PCR, revealing a significant decline of the relative expression of genes in RNA samples of disparate quality, while samples of similar, even poor integrity were found highly comparable. We discuss the consequences of these observations to minimize artifactual detection of false positive and negative differential expression due to RNA integrity differences, and propose a scheme for the development of a standard operational procedure, with optional registration of RNA integrity metrics in public repositories of gene expression data.'), ('754nln40', 'The ratio between proteins P27 and replicase of Cocksfoot mottle virus (CfMV) is regulated via a −1 programmed ribosomal frameshift (−1 PRF). A minimal frameshift signal with a slippery U UUA AAC heptamer and a downstream stem–loop structure was inserted into a dual reporter vector and directed −1 PRF with an efficiency of 14.4 ± 1.9% in yeast and 2.4 ± 0.7% in bacteria. P27-encoding CfMV sequence flanking the minimal frameshift signal caused ∼2-fold increase in the −1 PRF efficiencies both in yeast and in bacteria. In addition to the expected fusion proteins, termination products ending putatively at the frameshift site were found in yeast cells. We propose that the amount of premature translation termination from control mRNAs played a role in determining the calculated −1PRF efficiency. Co-expression of CfMV P27 with the dual reporter vector containing the minimal frameshift signal reduced the production of the downstream reporter, whereas replicase co-expression had no pronounced effect. This finding allows us to propose that CfMV protein P27 may influence translation at the frameshift site but the mechanism needs to be elucidated.'), ('p34ezktf', 'In Australia, compared with other developed countries the many and varied programs which comprise public health have continued to be funded poorly and unsystematically, particularly given the amount of publicly voiced political support. In 2003, the major public health policy developments in communicable disease control were in the fields of SARS, and vaccine funding, whilst the TGA was focused on the Pan Pharmaceutical crisis. Programs directed to health maintenance and healthy ageing were approved. The tertiary education sector was involved in the development of programs for training the public health workforce and new professional qualifications and competencies. The Abelson Report received support from overseas experts, providing a potential platform for calls to improve national funding for future Australian preventive programs; however, inconsistencies continued across all jurisdictions in their approaches to tackling national health priorities. Despite 2004 being an election year, public health policy was not visible, with the bulk of the public health funding available in the 2004/05 federal budget allocated to managing such emerging risks as avian flu. We conclude by suggesting several implications for the future.'), ('l3z27806', 'GIDEON (Global Infectious Diseases and Epidemiology Network) is a web-based computer program designed for decision support and informatics in the field of Geographic Medicine. The first of four interactive modules generates a ranked differential diagnosis based on patient signs, symptoms, exposure history and country of disease acquisition. Additional options include syndromic disease surveillance capability and simulation of bioterrorism scenarios. The second module accesses detailed and current information regarding the status of 338 individual diseases in each of 220 countries. Over 50,000 disease images, maps and user-designed graphs may be downloaded for use in teaching and preparation of written materials. The third module is a comprehensive source on the use of 328 anti-infective drugs and vaccines, including a listing of over 9,500 international trade names. The fourth module can be used to characterize or identify any bacterium or yeast, based on laboratory phenotype. GIDEON is an up-to-date and comprehensive resource for Geographic Medicine.'), ('yz2wbpuu', \"This debut editorial of Globalization and Health introduces the journal, briefly delineating its goals and objectives and outlines its scope of subject matter. 'Open Access' publishing is expected to become an increasingly important format for peer reviewed academic journals and that Globalization and Health is 'Open Access' is appropriate. The rationale behind starting a journal dedicated to globalization and health is three fold: Firstly: Globalization is reshaping the social geography within which we might strive to create health or prevent disease. The determinants of health – be they a SARS virus or a predilection for fatty foods – have joined us in our global mobility. Driven by economic liberalization and changing technologies, the phenomenon of 'access' is likely to dominate to an increasing extent the unfolding experience of human disease and wellbeing. Secondly: Understanding globalization as a subject matter itself needs certain benchmarks and barometers of its successes and failings. Health is one such barometer. It is a marker of social infrastructure and social welfare and as such can be used to either sound an alarm or give a victory cheer as our interconnectedness hurts and heals the populations we serve. And lastly: In as much as globalization can have an effect on health, it is also true that health and disease has an effect on globalization as exemplified by the existence of quarantine laws and the devastating economic effects of the AIDS pandemic. A balanced view would propose that the effects of globalization on health (and health systems) are neither universally good nor bad, but rather context specific. If the dialogue pertaining to globalization is to be directed or biased in any direction, then it must be this: that we consider the poor first.\"), ('kvhoa2se', \"In some Candida species, the universal CUG leucine codon is translated as serine. However, in most cases, the serine tRNAs responsible for this non-universal decoding (tRNA(Ser)CAG) accept in vitro not only serine, but also, to some extent, leucine. Nucleotide replacement experiments indicated that m1G37 is critical for leucylation activity. This finding was supported by the fact that the tRNA(Ser)CAGs possessing the leucylation activity always have m1G37, whereas that of Candida cylindracea, which possesses no leucylation activity, has A37. Quantification of defined aminoacetylated tRNAs in cells demonstrated that 3% of the tRNA(Ser)CAGs possessing m1G37 were, in fact, charged with leucine in vivo. A genetic approach using an auxotroph mutant of C.maltosa possessing this type of tRNA(Ser)CAG also suggested that the URA3 gene inactivated due to the translation of CUG as serine was rescued by a slight incorporation of leucine into the polypeptide, which demonstrated that the tRNA charged with multiple amino acids could participate in the translation. These findings provide the first evidence that two distinct amino acids are assigned by a single codon, which occurs naturally in the translation process of certain Candida species. We term this novel type of codon a 'polysemous codon'.\"), ('cgl34ykt', \"At present, there is no doubt that RNA recombination is one of the major factors responsible for the generation of new RNA viruses and retroviruses. Numerous experimental systems have been created to investigate this complex phenomenon. Consequently, specific RNA structural motifs mediating recombination have been identified in several viruses. Unfortunately, up till now a unified model of genetic RNA recombination has not been formulated, mainly due to difficulties with the direct comparison of data obtained for different RNA-based viruses. To solve this problem, we have attempted to construct a universal system in which the recombination activity of various RNA sequences could be tested. To this end, we have used brome mosaic virus, a model (+)RNA virus of plants, for which the structural requirements of RNA recombination are well defined. The effectiveness of the new homomolecular system has been proven in an experiment involving two RNA sequences derived from the hepatitis C virus genome. In addition, comparison of the data obtained with the homomolecular system with those generated earlier using the heteromolecular one has provided new evidence that the mechanisms of homologous and non-homologous recombination are different and depend on the virus' mode of replication.\"), ('ajlctjeb', 'BACKGROUND: Mammalian reoviruses naturally infect their hosts through the enteric and respiratory tracts. During enteric infections, proteolysis of the reovirus outer capsid protein σ3 is mediated by pancreatic serine proteases. In contrast, the proteases critical for reovirus replication in the lung are unknown. Neutrophil elastase (NE) is an acid-independent, inflammatory serine protease predominantly expressed by neutrophils. In addition to its normal role in microbial defense, aberrant expression of NE has been implicated in the pathology of acute respiratory distress syndrome (ARDS). Because reovirus replication in rodent lungs causes ARDS-like symptoms and induces an infiltration of neutrophils, we investigated the capacity of NE to promote reovirus virion uncoating. RESULTS: The human promonocyte cell line U937 expresses NE. Treatment of U937 cells with the broad-spectrum cysteine-protease inhibitor E64 [trans-epoxysuccinyl-L-leucylamido-(4-guanidino)butane] and with agents that increase vesicular pH did not inhibit reovirus replication. Even when these inhibitors were used in combination, reovirus replicated to significant yields, indicating that an acid-independent non-cysteine protease was capable of mediating reovirus uncoating in U937 cell cultures. To identify the protease(s) responsible, U937 cells were treated with phorbol 12-myristate 13-acetate (PMA), an agent that induces cellular differentiation and results in decreased expression of acid-independent serine proteases, including NE and cathepsin (Cat) G. In the presence of E64, reovirus did not replicate efficiently in PMA-treated cells. To directly assess the role of NE in reovirus infection of U937 cells, we examined viral growth in the presence of N-Ala-Ala-Pro-Val chloromethylketone, a NE-specific inhibitor. Reovirus replication in the presence of E64 was significantly reduced by treatment of cells with the NE inhibitor. Incubation of virions with purified NE resulted in the generation of infectious subviron particles that did not require additional intracellular proteolysis. CONCLUSION: Our findings reveal that NE can facilitate reovirus infection. The fact that it does so in the presence of agents that raise vesicular pH supports a model in which the requirement for acidic pH during infection reflects the conditions required for optimal protease activity. The capacity of reovirus to exploit NE may impact viral replication in the lung and other tissues during natural infections.'), ('cl9gpt9w', 'The influence of locked nucleic acid (LNA) residues on the thermodynamic properties of 2′-O-methyl RNA/RNA heteroduplexes is reported. Optical melting studies indicate that LNA incorporated into an otherwise 2′-O-methyl RNA oligonucleotide usually, but not always, enhances the stabilities of complementary duplexes formed with RNA. Several trends are apparent, including: (i) a 3′ terminal U LNA and 5′ terminal LNAs are less stabilizing than interior and other 3′ terminal LNAs; (ii) most of the stability enhancement is achieved when LNA nucleotides are separated by at least one 2′-O-methyl nucleotide; and (iii) the effects of LNA substitutions are approximately additive when the LNA nucleotides are separated by at least one 2′-O-methyl nucleotide. An equation is proposed to approximate the stabilities of complementary duplexes formed with RNA when at least one 2′-O-methyl nucleotide separates LNA nucleotides. The sequence dependence of 2′-O-methyl RNA/RNA duplexes appears to be similar to that of RNA/RNA duplexes, and preliminary nearest-neighbor free energy increments at 37°C are presented for 2′-O-methyl RNA/RNA duplexes. Internal mismatches with LNA nucleotides significantly destabilize duplexes with RNA.'), ('t40ybhgb', 'Sequencing pathogen genomes is costly, demanding careful allocation of limited sequencing resources. We built a computational Sequencing Analysis Pipeline (SAP) to guide decisions regarding the amount of genomic sequencing necessary to develop high-quality diagnostic DNA and protein signatures. SAP uses simulations to estimate the number of target genomes and close phylogenetic relatives (near neighbors or NNs) to sequence. We use SAP to assess whether draft data are sufficient or finished sequencing is required using Marburg and variola virus sequences. Simulations indicate that intermediate to high-quality draft with error rates of 10(−3)–10(−5) (∼8× coverage) of target organisms is suitable for DNA signature prediction. Low-quality draft with error rates of ∼1% (3× to 6× coverage) of target isolates is inadequate for DNA signature prediction, although low-quality draft of NNs is sufficient, as long as the target genomes are of high quality. For protein signature prediction, sequencing errors in target genomes substantially reduce the detection of amino acid sequence conservation, even if the draft is of high quality. In summary, high-quality draft of target and low-quality draft of NNs appears to be a cost-effective investment for DNA signature prediction, but may lead to underestimation of predicted protein signatures.'), ('zwbc7nnn', \"BACKGROUND: Epitopes can be defined as the molecular structures bound by specific receptors, which are recognized during immune responses. The Immune Epitope Database and Analysis Resource (IEDB) project will catalog and organize information regarding antibody and T cell epitopes from infectious pathogens, experimental antigens and self-antigens, with a priority on NIAID Category A-C pathogens () and emerging/re-emerging infectious diseases. Both intrinsic structural and phylogenetic features, as well as information relating to the interactions of the epitopes with the host's immune system will be catalogued. DESCRIPTION: To effectively represent and communicate the information related to immune epitopes, a formal ontology was developed. The semantics of the epitope domain and related concepts were captured as a hierarchy of classes, which represent the general and specialized relationships between the various concepts. A complete listing of classes and their properties can be found at . CONCLUSION: The IEDB's ontology is the first ontology specifically designed to capture both intrinsic chemical and biochemical information relating to immune epitopes with information relating to the interaction of these structures with molecules derived from the host immune system. We anticipate that the development of this type of ontology and associated databases will facilitate rigorous description of data related to immune epitopes, and might ultimately lead to completely new methods for describing and modeling immune responses.\"), ('1r65yam5', 'BACKGROUND: Salmonid fishes are among the most widely studied model fish species but reports on systematic evaluation of reference genes in qRT-PCR studies is lacking. RESULTS: The stability of six potential reference genes was examined in eight tissues of Atlantic salmon (Salmo salar), to determine the most suitable genes to be used in quantitative real-time RT-PCR analyses. The relative transcription levels of genes encoding 18S rRNA, S20 ribosomal protein, β-actin, glyceraldehyde-3P-dehydrogenase (GAPDH), and two paralog genes encoding elongation factor 1A (EF1A(A )and EF1A(B)) were quantified in gills, liver, head kidney, spleen, thymus, brain, muscle, and posterior intestine in six untreated adult fish, in addition to a group of individuals that went through smoltification. Based on calculations performed with the geNorm VBA applet, which determines the most stable genes from a set of tested genes in a given cDNA sample, the ranking of the examined genes in adult Atlantic salmon was EF1A(B)>EF1A(A)>β-actin>18S rRNA>S20>GAPDH. When the same calculations were done on a total of 24 individuals from four stages in the smoltification process (presmolt, smolt, smoltified seawater and desmoltified freshwater), the gene ranking was EF1A(B)>EF1A(A)>S20>β-actin>18S rRNA>GAPDH. CONCLUSION: Overall, this work suggests that the EF1A(A )and EF1A(B )genes can be useful as reference genes in qRT-PCR examination of gene expression in the Atlantic salmon.'), ('oa4lzkru', 'Zoonoses, diseases that jump from animals to humans, are a growing health problem around the world. Understanding their causes and their effects on humans have therefore become an important topic for global public health'), ('qva0jt86', 'BACKGROUND AND METHODS: Human metapneumovirus (hMPV) is a recently discovered respiratory virus associated with bronchiolitis, pneumonia, croup and exacerbations of asthma. Since respiratory viruses are frequently detected in patients with acute exacerbations of COPD (AE-COPD) it was our aim to investigate the frequency of hMPV detection in a prospective cohort of hospitalized patients with AE-COPD compared to patients with stable COPD and to smokers without by means of quantitative real-time RT-PCR. RESULTS: We analysed nasal lavage and induced sputum of 130 patients with AE-COPD, 65 patients with stable COPD and 34 smokers without COPD. HMPV was detected in 3/130 (2.3%) AE-COPD patients with a mean of 6.5 × 10(5 )viral copies/ml in nasal lavage and 1.88 × 10(5 )viral copies/ml in induced sputum. It was not found in patients with stable COPD or smokers without COPD. CONCLUSION: HMPV is only found in a very small number of patients with AE-COPD. However it should be considered as a further possible viral trigger of AE-COPD because asymptomatic carriage is unlikely.'), ('vw8xjo9t', 'The BIG project looks at some of the ethical concerns surrounding globalization and health.'), ('bnnl700a', 'BACKGROUND: The present study aimed to provide information on awareness of the attributable fraction of cancer causes among the Japanese general population. METHODS: A nationwide representative sample of 2,000 Japanese aged 20 or older was asked about their perception and level of concern about various environmental and genetic risk factors in relation to cancer prevention, as a part of an Omnibus Survey. Interviews were conducted with 1,355 subjects (609 men and 746 women). RESULTS: Among 12 risk factor candidates, the attributable fraction of cancer-causing viral and bacterial infection was considered highest (51%), followed by that of tobacco smoking (43%), stress (39%), and endocrine-disrupting chemicals (37%). On the other hand, the attributable fractions of cancer by charred fish and meat (21%) and alcohol drinking (22%) were considered low compared with other risk factor candidates. For most risk factors, attributable fraction responses were higher in women than in men. As a whole, the subjects tended to respond with higher values than those estimated by epidemiologic evidence in the West. The attributable fraction of cancer speculated to be genetically determined was 32%, while 36% of cancer was considered preventable by improving lifestyle. CONCLUSION: Our results suggest that awareness of the attributable fraction of cancer causes in the Japanese general population tends to be dominated by cancer-causing infection, occupational exposure, air pollution and food additives rather than major lifestyle factors such as diet.'), ('m71xkuo9', 'BACKGROUND: Development of a practical gene point-of-care testing device (g-POCT device) requires innovative detection methods for demonstrating the results of the gene amplification reaction without the use of expensive equipment. We have studied a new method for the sequence-specific visual detection of minute amounts of nucleic acids using precipitation reaction by addition of cationic polymers to amplicons of Loop mediated isothermal Amplification (LAMP). RESULTS: Oligo DNA probes labeled with different fluorescent dyes were prepared for multiple nucleic acid templates, and the templates were amplified by the LAMP reactions under the existence of the probes. At completion of the LAMP reaction, an optimal amount of low molecular weight polyethylenimine (PEI) was added, resulting in the precipitation of the insoluble LAMP amplicon-PEI complex. The fluorescently labeled Oligo DNA probes hybridized to the LAMP product were incorporated into the precipitation, and the precipitate emitted fluorescence corresponding to the amplified nucleic acid templates. The color of emitted fluorescence can be detected easily by naked eye on a conventional UV illuminator. CONCLUSION: The presence or absence of minute amount of nucleic acid templates could be detected in a simple manner through visual assessment for the color of the LAMP amplicon-PEI complex precipitate. We conclude that this detection method may facilitate development of small and simple g-POCT device.'), ('s64v656n', 'Illicit drug abuse and HIV/AIDS have increased rapidly in the past 10 to 20 years in China. This paper reviews drug abuse in China, the HIV/AIDS epidemic and its association with injection drug use (IDU), and Chinese policies on illicit drug abuse and prevention of HIV/AIDS based on published literature and unpublished official data. As a major drug trans-shipment country with source drugs from the \"Golden Triangle\" and \"Gold Crescent\" areas in Asia, China has also become an increasingly important drug consuming market. About half of China\\'s 1.14 million documented drug users inject, and many share needles. IDU has contributed to 42% of cumulatively reported HIV/AIDS cases thus far. Drug trafficking is illegal in China and can lead to the death penalty. The public security departments adopt \"zero tolerance\" approach to drug use, which conflict with harm reduction policies of the public health departments. Past experience in China suggests that cracking down on drug smuggling and prohibiting drug use alone can not prevent or solve all illicit drug related problems in the era of globalization. In recent years, the central government has outlined a series of pragmatic policies to encourage harm reduction programs; meanwhile, some local governments have not fully mobilized to deal with drug abuse and HIV/AIDS problems seriously. Strengthening government leadership at both central and local levels; scaling up methadone substitution and needle exchange programs; making HIV voluntary counseling and testing available and affordable to both urban and rural drug users; and increasing utilization of outreach and nongovernmental organizations are offered as additional strategies to help cope with China\\'s HIV and drug abuse problem.'), ('oluq7v0h', 'Neutralization of West Nile virus (WNV) in vivo correlates with the development of an antibody response against the viral envelope (E) protein. Using random mutagenesis and yeast surface display, we defined individual contact residues of 14 newly generated monoclonal antibodies against domain III of the WNV E protein. Monoclonal antibodies that strongly neutralized WNV localized to a surface patch on the lateral face of domain III. Convalescent antibodies from individuals who had recovered from WNV infection also detected this epitope. One monoclonal antibody, E16, neutralized 10 different strains in vitro, and showed therapeutic efficacy in mice, even when administered as a single dose 5 d after infection. A humanized version of E16 was generated that retained antigen specificity, avidity and neutralizing activity. In postexposure therapeutic trials in mice, a single dose of humanized E16 protected mice against WNV-induced mortality, and may therefore be a viable treatment option against WNV infection in humans. SUPPLEMENTARY INFORMATION: The online version of this article (doi:10.1038/nm1240) contains supplementary material, which is available to authorized users.'), ('tw6wusxe', 'BACKGROUND: Current national preparedness plans require local health departments to play an integral role in responding to an influenza pandemic, a major public health threat that the World Health Organization has described as \"inevitable and possibly imminent\". To understand local public health workers\\' perceptions toward pandemic influenza response, we surveyed 308 employees at three health departments in Maryland from March – July 2005, on factors that may influence their ability and willingness to report to duty in such an event. RESULTS: The data suggest that nearly half of the local health department workers are likely not to report to duty during a pandemic. The stated likelihood of reporting to duty was significantly greater for clinical (Multivariate OR: 2.5; CI 1.3–4.7) than technical and support staff, and perception of the importance of one\\'s role in the agency\\'s overall response was the single most influential factor associated with willingness to report (Multivariate OR: 9.5; CI 4.6–19.9). CONCLUSION: The perceived risk among public health workers was shown to be associated with several factors peripheral to the actual hazard of this event. These risk perception modifiers and the knowledge gaps identified serve as barriers to pandemic influenza response and must be specifically addressed to enable effective local public health response to this significant threat.'), ('58czem0j', 'BACKGROUND: As a number of commentators have noted, SARS exposed the vulnerabilities of our health care systems and governance structures. Health care professionals (HCPs) and hospital systems that bore the brunt of the SARS outbreak continue to struggle with the aftermath of the crisis. Indeed, HCPs – both in clinical care and in public health – were severely tested by SARS. Unprecedented demands were placed on their skills and expertise, and their personal commitment to their profession was severely tried. Many were exposed to serious risk of morbidity and mortality, as evidenced by the World Health Organization figures showing that approximately 30% of reported cases were among HCPs, some of whom died from the infection. Despite this challenge, professional codes of ethics are silent on the issue of duty to care during communicable disease outbreaks, thus providing no guidance on what is expected of HCPs or how they ought to approach their duty to care in the face of risk. DISCUSSION: In the aftermath of SARS and with the spectre of a pandemic avian influenza, it is imperative that we (re)consider the obligations of HCPs for patients with severe infectious diseases, particularly diseases that pose risks to those providing care. It is of pressing importance that organizations representing HCPs give clear indication of what standard of care is expected of their members in the event of a pandemic. In this paper, we address the issue of special obligations of HCPs during an infectious disease outbreak. We argue that there is a pressing need to clarify the rights and responsibilities of HCPs in the current context of pandemic flu preparedness, and that these rights and responsibilities ought to be codified in professional codes of ethics. Finally, we present a brief historical accounting of the treatment of the duty to care in professional health care codes of ethics. SUMMARY: An honest and critical examination of the role of HCPs during communicable disease outbreaks is needed in order to provide guidelines regarding professional rights and responsibilities, as well as ethical duties and obligations. With this paper, we hope to open the social dialogue and advance the public debate on this increasingly urgent issue.'), ('x7wg7e9s', 'The anticancer potential of catechins derived from green tea is not well understood, in part because catechin-related growth suppression and/or apoptosis appears to vary with the type and stage of malignancy as well as with the type of catechin. This in vitro study examined the biological effects of epicatechin (EC), epigallocatechin (EGC), EC 3-gallate (ECG) and EGC 3-gallate (EGCG) in cell lines from human gender-specific cancers. Cell lines developed from organ-confined (HH870) and metastatic (DU145) prostate cancer, and from moderately (HH450) and poorly differentiated (HH639) epithelial ovarian cancer were grown with or without EC, EGC, ECG or EGCG. When untreated cells reached confluency, viability and doubling time were measured for treated and untreated cells. Whereas EC treatment reduced proliferation of HH639 cells by 50%, EGCG suppressed proliferation of all cell lines by 50%. ECG was even more potent: it inhibited DU145, HH870, HH450 and HH639 cells at concentrations of 24, 27, 29 and 30 µM, whereas EGCG inhibited DU145, HH870, HH450 and HH639 cells at concentrations 89, 45, 62 and 42 µM. When compared with EGCG, ECG more effectively suppresses the growth of prostate cancer and epithelial ovarian cancer cell lines derived from tumors of patients with different stages of disease.'), ('utglk4af', \"BACKGROUND: Patients with chronic obstructive pulmonary disease (COPD) can experience 'exacerbations' of their conditions. An exacerbation is an event defined in terms of subjective descriptors or symptoms, namely dyspnoea, cough and sputum that worsen sufficiently to warrant a change in medical management. There is a need for reliable markers that reflect the pathological mechanisms that underlie exacerbation severity and that can be used as a surrogate to assess treatment effects in clinical studies. Little is known as to how existing study variables and suggested markers change in both the stable and exacerbation phases of COPD. In an attempt to find the best surrogates for exacerbations, we have reviewed the literature to identify which of these markers change in a consistent manner with the severity of the exacerbation event. METHODS: We have searched standard databases between 1966 to July 2004 using major keywords and terms. Studies that provided demographics, spirometry, potential markers, and clear eligibility criteria were included in this study. Central tendencies and dispersions for all the variables and markers reported and collected by us were first tabulated according to sample size and ATS/ERS 2004 Exacerbation Severity Levels I to III criteria. Due to the possible similarity of patients in Levels II and III, the data was also redefined into categories of exacerbations, namely out-patient (Level I) and in-patient (Levels II & III combined). For both approaches, we performed a fixed effect meta-analysis on each of the reported variables. RESULTS: We included a total of 268 studies reported between 1979 to July 2004. These studies investigated 142,407 patients with COPD. Arterial carbon dioxide tension and breathing rate were statistically different between all levels of exacerbation severity and between in out- and in-patient settings. Most other measures showed weak relationships with either level or setting, or they had insufficient data to permit meta-analysis. CONCLUSION: Arterial carbon dioxide and breathing rate varied in a consistent manner with exacerbation severity and patient setting. Many other measures showed weak correlations that should be further explored in future longitudinal studies or assessed using suggested mathematical modelling techniques.\"), ('i5fcedbo', ''), ('bjjft7ut', ''), ('urk7fe34', ''), ('u09rn09u', ''), ('kvztcwu2', ''), ('9r62ffew', 'BACKGROUND: The main limitation in performing genome-wide gene-expression profiling is the assay of low-expression genes. Approaches with high throughput and high sensitivity for assaying low-expression transcripts are urgently needed for functional genomic studies. Combination of the suppressive subtractive hybridization (SSH) and cDNA microarray techniques using the subtracted cDNA clones as probes printed on chips has greatly improved the efficiency for fishing out the differentially expressed clones and has been used before. However, it remains tedious and inefficient sequencing works for identifying genes including the great number of redundancy in the subtracted amplicons, and sacrifices the original advantages of high sensitivity of SSH in profiling low-expression transcriptomes. RESULTS: We modified the previous combination of SSH and microarray methods by directly using the subtracted amplicons as targets to hybridize the pre-made cDNA microarrays (named as \"SSH/microarray\"). mRNA prepared from three pairs of hepatoma and non-hepatoma liver tissues was subjected to the SSH/microarray assays, as well as directly to regular cDNA microarray assays for comparison. As compared to the original SSH and microarray combination assays, the modified SSH/microarray assays allowed for much easier inspection of the subtraction efficiency and identification of genes in the subtracted amplicons without tedious and inefficient sequencing work. On the other hand, 5015 of the 9376 genes originally filtered out by the regular cDNA microarray assays because of low expression became analyzable by the SSH/microarray assays. Moreover, the SSH/microarray assays detected about ten times more (701 vs. 69) HCC differentially expressed genes (at least a two-fold difference and P < 0.01), particularly for those with rare transcripts, than did the regular cDNA microarray assays. The differential expression was validated in 9 randomly selected genes in 18 pairs of hepatoma/non-hepatoma liver tissues using quantitative RT-PCR. The SSH/microarray approaches resulted in identifying many differentially expressed genes implicated in the regulation of cell cycle, cell death, signal transduction and cell morphogenesis, suggesting the involvement of multi-biological processes in hepato-carcinogenesis. CONCLUSION: The modified SSH/microarray approach is a simple but high-sensitive and high-efficient tool for differentially profiling the low-expression transcriptomes. It is most adequate for applying to functional genomic studies.'), ('d3cko4j2', 'The development of DNA microarray technology a decade ago led to the establishment of functional genomics as one of the most active and successful scientific disciplines today. With the ongoing development of immunomic microarray technology—a spatially addressable, large-scale technology for measurement of specific immunological response—the new challenge of functional immunomics is emerging, which bears similarities to but is also significantly different from functional genomics. Immunonic data has been successfully used to identify biological markers involved in autoimmune diseases, allergies, viral infections such as human immunodeficiency virus (HIV), influenza, diabetes, and responses to cancer vaccines. This review intends to provide a coherent vision of this nascent scientific field, and speculate on future research directions. We discuss at some length issues such as epitope prediction, immunomic microarray technology and its applications, and computation and statistical challenges related to functional immunomics. Based on the recent discovery of regulation mechanisms in T cell responses, we envision the use of immunomic microarrays as a tool for advances in systems biology of cellular immune responses, by means of immunomic regulatory network models.')]\n"
     ]
    }
   ],
   "source": [
    "from pyspark.sql import SparkSession\n",
    "from pyspark.ml.feature import HashingTF, IDF\n",
    "from pyspark.ml.linalg import SparseVector, DenseVector\n",
    "from itertools import combinations\n",
    "import json\n",
    "\n",
    "\n",
    "def load_documents(spark, input_path):\n",
    "    \"\"\"\n",
    "    Carica i documenti dal file JSONL e restituisce un RDD di Spark.\n",
    "    Il file JSONL deve contenere un documento per riga.\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    # Leggi il file JSONL come RDD di Spark\n",
    "    rdd = spark.sparkContext.textFile(input_path)\n",
    "\n",
    "    \n",
    "    # Estrai l'id e il testo di ogni documento\n",
    "    documents = rdd.map(lambda line: json.loads(line)).map(lambda d: (d[\"_id\"], d[\"text\"]))\n",
    "    return documents\n",
    "\n",
    "documents = load_documents(spark, path_name_documents)\n",
    "documents_list = documents.collect()\n",
    "print(documents_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdba7c51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fa506bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark.sql import SparkSession\n",
    "from pyspark.ml.feature import HashingTF, IDF\n",
    "from pyspark.ml.linalg import SparseVector, DenseVector\n",
    "from itertools import combinations\n",
    "import json\n",
    "\n",
    "\n",
    "def preprocess_document(document, hashingTF, idfModel):\n",
    "    \"\"\"\n",
    "    Applica la pre-elaborazione a un singolo documento, includendo la rappresentazione HashingTF\n",
    "    e la rappresentazione IDF.\n",
    "    Restituisce un oggetto DenseVector che rappresenta il documento.\n",
    "    \"\"\"\n",
    "\n",
    "    # Estra il testo dal documento\n",
    "    text = document[1]\n",
    "\n",
    "    # Crea una lista di tuple (indice, valore) per ogni parola nel testo\n",
    "    word_counts = hashingTF.transform(text.split())\n",
    "    word_tfidf = idfModel.transform(word_counts)\n",
    "    words = [(i, v) for i, v in enumerate(word_tfidf.toArray()) if v != 0.0]\n",
    "\n",
    "    # Crea un oggetto SparseVector che rappresenta il documento\n",
    "    if len(words) == 0:\n",
    "        return None\n",
    "    else:\n",
    "        return DenseVector([v for i, v in words])\n",
    "\n",
    "def calculate_similarity(documents):\n",
    "    \"\"\"\n",
    "    Calcola la cosine similarity tra tutte le coppie di documenti.\n",
    "    Restituisce un RDD di tuple (id_doc_1, id_doc_2, cosine_similarity).\n",
    "    \"\"\"\n",
    "\n",
    "    # Calcola la cosine similarity tra tutte le coppie di documenti\n",
    "    document_pairs = combinations(documents.collect(), 2)\n",
    "    similarities = []\n",
    "    for doc1, doc2 in document_pairs:\n",
    "        if doc1 is not None and doc2 is not None:\n",
    "            cosine_sim = doc1.dot(doc2) / (doc1.norm(2) * doc2.norm(2))\n",
    "            similarities.append((doc1[0], doc2[0], cosine_sim))\n",
    "\n",
    "    return similarities\n",
    "\n",
    "\n",
    "def main():\n",
    "\n",
    "    # Carica i documenti dal file JSONL\n",
    "    documents = load_documents(spark, path_name_documents)\n",
    "\n",
    "    # Crea un oggetto HashingTF per la rappresentazione dei documenti come bag of words\n",
    "    hashingTF = HashingTF(numFeatures=10000, inputCol=\"text\", outputCol=\"word_count\")\n",
    "\n",
    "    # Calcola la rappresentazione IDF dei documenti\n",
    "    word_counts = hashingTF.transform(documents)\n",
    "    idf = IDF(inputCol=\"word_count\", outputCol=\"features\")\n",
    "    idfModel = idf.fit(word_counts)\n",
    "\n",
    "    # Pre-elabora tutti i documenti\n",
    "    preprocessed_documents = documents.map(lambda d: (d[0], preprocess_document(d, hashingTF, idfModel))) \\\n",
    "                                       .filter(lambda d: d[1] is not None)\n",
    "\n",
    "    # Calcola la cosine similarity tra tutte le coppie di documenti\n",
    "    similarities = calculate_similarity(preprocessed_documents)\n",
    "\n",
    "\n",
    "\n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ed0072f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f6f7bd5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a68ffaa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyspark import SparkConf\n",
    "from pyspark import SparkContext\n",
    "\n",
    "import os\n",
    "\n",
    "os.environ['PYSPARK_PYTHON'] = 'C:/Users/lita4/anaconda3/python.exe'\n",
    "os.environ['PYSPARK_DRIVER_PYTHON'] = 'C:/Users/lita4/anaconda3/python.exe'\n",
    "\n",
    "conf = SparkConf().setAppName(\"MyApp\")\n",
    "sc = SparkContext(conf=conf)\n",
    "\n",
    "\n",
    "def sq(x):\n",
    "    return x*x\n",
    "\n",
    "data = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]\n",
    "myrdd = sc.parallelize(data)\n",
    "squared = myrdd.map(sq) # this creates a new RDD\n",
    "print(squared.collect()) # careful: collect() is an action"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3fefe51",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
